---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', fig.width=9)
```

# Overview

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(gtools)
project_summary_mm10 = "/home/lpantano/ody/projects/nytia_singlecell/singlecell/final/2014-10-08_singlecell/project-summar-2y.csv"
counts_file = "/home/lpantano/ody/projects/nytia_singlecell/singlecell3/final/2014-10-22_singlecell3/combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = read.table(project_summary, header=TRUE, sep=",")
rownames(summarydata) = summarydata$Name
summarydata$Name <- factor(summarydata$Name, levels=mixedsort(summarydata$Name))
summarydata = summarydata[order(rownames(summarydata)),]
counts = read.table(counts_file, header=TRUE, row.names="id")
counts = counts[, order(colnames(counts))]
```

```{r path-info}
source("~/repos/myRfunctions/transactions.R")
dn <- "nitya_singlecell"
root_path<-"~/repos/pipelines/nitya_singlecell"
setwd(root_path)
```

```{r render,eval=FALSE}
render_2_drop("~/repos/pipelines/nitya_singlecell/qc-summary-human-alone.Rmd",dn)
```


# Quality control metrics

## Mapped reads
```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped, fill = type)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapped reads") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```


## Genomic mapping rate
```{r mapping-rate-plot}

ggplot(summarydata, aes(x=Name, y=Mapping.Rate, fill = type)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```


## Number of genes detected
```{r genes-detected-plot}
ggplot(summarydata, aes(x=Name, y=Genes.Detected, fill = type)) +
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) + 
    ylab("genes detected") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Exonic mapping rate
```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate, fill = type)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("exonic mapping rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate, fill = type)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("rRNA rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Estimated fragment length of paired-end reads
```{r fragment-length-plot}
ggplot(summarydata, aes(x=Name, y=Fragment.Length.Mean, fill = type)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("fragment length") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Correlation heatmap of raw counts
```{r heatmap-raw}
p = ggheatmap(cor(counts)) 
ggheatmap.show(p)
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Correlation heatmap of TMM-normalized counts
```{r heatmap-normalized}
p = ggheatmap(cor(normalized_counts))
ggheatmap.show(p)
```

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized}
p = ggheatmap(cor(normalized_counts, method="spearman"))
ggheatmap.show(p)
```

## MDS plot of TMM-normalized counts
```{r mds-normalized}
mds(normalized_counts, k=length(colnames(normalized_counts)) - 1)
```

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select = order(rowMeans(counts),decreasing=TRUE)[1:30]
p = ggheatmap(as.matrix(counts[select,]))
ggheatmap.show(p)
```
