---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

# Report

### Setup: loading packages and data

```{r setup}
library(knitr)
library(rmarkdown)
library(knitrBootstrap)
library(ggplot2)
library(reshape)
library(DESeq2)
library(genefilter)
library(CHBUtils)
library(gtools)
library(gridExtra)
library(devtools)
library(dplyr)
devtools::load_all("~/repos/isomiRs")
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
source("~/repos/myRfunctions/transactions.R")
dn <- "tina_exosomes"
root_path<-"~/orch/scratch/tina_exosomes"
```

```{r render,eval=FALSE}
render_2_drop("~/repos/pipelines/tina_exosomes/report/mirna.rmd",dn)
```


```{r load}
setwd(root_path)
files <- read.table("mirna.files")
files <- files[1:5,]
files$V1 <- paste0("mirna/",files$V1)
row.names(files) <- files[,2]
design <- data.frame(type=files[,2],row.names=files[,2])
obj <- loadIso(files = files[,1], design = design ,header = T)

# clus <- read.table("cluster/res/counts.tsv",header=T,sep="\t",row.names=1)
# names(clus) <- gsub("mir.71","mir71",names(clus))
# ann <- as.vector(unlist(as.data.frame((sapply(as.character(clus[,1]),strsplit, split="\\|")))[1,]))
# clus_ma <- clus[,2:ncol(clus)]
# clus_ma <- clus_ma[,row.names(design)]

# total_clus <- read.table("cluster/res/stats_align.dat")
# names(total_clus) <- c("size","sample","reads")
# total_clus$sample <- gsub("mir-71","mir71",total_clus$sample)
```



## small RNA size distribution

### After adapter recognition

I removed one sample because it has only 500 reads: `Cell_S3`

```{r adapter}
ad_recog<-data.frame()
total_reads<-data.frame()
for (nr in 1:nrow(files)){
    ad <- read.table(paste0( root_path,"/",gsub("-ann.mirna",".info",files[nr,1]) ),
                     skip=12,sep=" ")
    ad_recog <- rbind(ad_recog, data.frame( pos=ad$V2,
                                            value=ad$V6,
                                            sample=files[nr,2]                                                                                
                                            ))
    ad <- read.table(paste0( root_path,"/",gsub("-ann.mirna",".info",files[nr,1]) ),nrows = 1,sep=":")
    total_reads = rbind(total_reads, data.frame(sample = files[nr,2], totalreads=ad[1,2]))
}

ggplot(ad_recog)+
    geom_bar(aes(x=pos,y=value,fill=sample),
             stat='identity',
             position='dodge')+
    xlim(c(0,50)) +
    labs(list(x="size",title="size distribution",y="reads"))

```

Here we can see how much of the total reads has adapter

* `mirna_ratio` is calculated over the total reads with adapter

```{r adapter-step,results='asis'}
adapter_step = cbind(ad_recog %>% group_by(sample) %>% summarize(total=sum(value)),total_reads[c(4:5,1:3),])
adapter_step$w_adapter_ratio = adapter_step$total/adapter_step$totalreads*100
names(adapter_step)[2] = "w_adapter"
mirna_step <- as.data.frame(colSums(obj@counts))
counts_stats = cbind(adapter_step,mirna_step)
names(counts_stats)[6] = "mirna"
counts_stats$mirna_ratio = counts_stats$mirna/counts_stats$w_adapter*100
knitr::kable(counts_stats[,c(1,4,2,5,6,7)])
```


The parent cells have very low percentage of small RNA (reads with adapter)...
this could mean that we have mRNA there, and there is no enrichment in small RNA.
But the small RNA that there is, is enriched in 22 nt long, meaning miRNA, so that is good.
The exosomes samples are not enriched in miRNA, but we have a peak at 33. I would
say these fragments are tRNA, but need to proceed with the analysis to confirm. 
The bad thin is there is no 22 nt long enrichmend, so no miRNA enrichment.


## small RNA annotation

Mainly tRNA in exosomes

```{r clusters}
dd = read.table("~/orch/scratch/tina_exosomes/work/res/stats/stats_align.dat")
head(dd)
dd$ann = "Others"
dd$ann[grepl("tRNA", dd$V4)] = "tRNA"
dd$ann[grepl("miRNA", dd$V4)] = "miRNA"
dd$ann[dd$V4=="align"] = "align"
dd$ann[dd$V4=="json"] = "cluster"
bytype = dd %>% group_by(V1,V2,ann) %>% summarise(total=sum(V3))
ggplot(bytype, aes(y=total,x=V1,fill=ann))+
    geom_bar(stat = "identity", position = "dodge")+
    facet_wrap(~V2)
```


## fragmens

The bigger fragments don't seem to be transcriptomic. There is big ammount that are unkown.

```{r rna}
dd = read.table("~/orch/scratch/tina_exosomes/work/exosomes_large/work/project-summary.csv", header=1,sep=",",row.names=1)
head(dd)
knitr::kable(dd[,c(11,12,8,2)])
```


# Methods

## mirnas

I used [miranalizer](https://github.com/lpantano/seqbuster/wiki/miraligner) to get the mirnas. The mainly different with mirdeep is that it
can take a look at isomirs. We can add this to your analysis, if you are interestd.
I used [isomiR](https://github.com/lpantano/isomiRs) package to manipulate the output of miranalizer.


## other small RNA

I used [seqcluster](https://github.com/lpantano/seqcluster), that can generate clusters of sequences over the genome, and 
annotate them to some type of small RNA, like rRNA, tRNA ... The main advantage is that
will count sequences that map in multi-places only once, so you can use the 
standard packages to analyze this data. We can add DE of other smallRNAs after we 
discuss these results. The mapping was done with bowtie1 in this case.
