---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---
```{r inside}
library(knitr)
library(rmarkdown)
library(knitrBootstrap)
source("~/repos/myRfunctions/transactions.R")
dn <- "mmu_p53"
root_path<-"~/orch/scratch/glasgow/glasgow/mmu_p53_ko/mmu_p53/res"
path_files = "~/repos/pipelines/glasgow/mmu_p53/"
```

```{r render,eval=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
                      cache.path="~/.cache/glasgow/html_cached/", eval=TRUE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
render_2_drop("~/repos/pipelines/glasgow/mmu_p53/qc-summary.Rmd",dn)
```


# Overview

```{r qc-setup}
library(ggplot2)
library(grid)
library(gridExtra)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
project_summary = "/home/lpantano/orch/scratch/glasgow/glasgow/mmu_p53_ko/mmu_p53/work/project-summary.csv"
counts_file = "/home/lpantano/orch/scratch/glasgow/glasgow/mmu_p53_ko/mmu_p53/res/2014-11-27_mmu_p53/combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = read.table(project_summary, header=TRUE, sep=",")
rownames(summarydata) = summarydata$Name
summarydata = summarydata[order(rownames(summarydata)),]
counts = read.table(counts_file, header=TRUE, row.names="id")
counts = counts[, order(colnames(counts))]
```

# Quality control metrics

## Mapped reads
```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapped reads") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Genomic mapping rate
```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Number of genes detected
```{r genes-detected-plot}
ggplot(summarydata, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) + 
    ylab("genes detected") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Exonic mapping rate
```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("exonic mapping rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("rRNA rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Estimated fragment length of paired-end reads
```{r fragment-length-plot}
ggplot(summarydata, aes(x=Name, y=Fragment.Length.Mean)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("fragment length") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Correlation heatmap of raw counts
```{r heatmap-raw}
p = ggheatmap(cor(counts)) 
ggheatmap.show(p)
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE, log = T)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Correlation heatmap of TMM-normalized counts
```{r heatmap-normalized}
p = ggheatmap(cor(normalized_counts))
ggheatmap.show(p)
```

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized}
p = ggheatmap(cor(normalized_counts, method="spearman"))
ggheatmap.show(p)
```

## MDS plot of TMM-normalized counts
```{r mds-normalized}
mds(normalized_counts, k=length(colnames(normalized_counts)) - 1)
```

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select = order(rowMeans(counts),decreasing=TRUE)[1:30]
p = ggheatmap(as.matrix(counts[select,]))
ggheatmap.show(p)
```

# Differential expression

```{r de-setup}
library(DESeq2)
library(DEGreport)
library(vsn)
summarydata$type = factor(c("Ctrl", "Ctrl","Ctrl", "p53", "p53", "p53"),levels=c("Ctrl","p53"))

design = ~ type
condition = "type"
dataset = "mmusculus_gene_ensembl"
filter = "ensembl_gene_id"
symbol = "external_gene_name"
```


```{r deseq2-expression-analysis, results='asis'}
counts <- counts[rowSums(counts>0)>1,]
dds = DESeqDataSetFromMatrix(countData=counts,
    colData=summarydata, design = design)
dds = DESeq(dds)
```

## Effect of variance stabilization
```{r deseq-diagnostics, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

## Dispersion estimates
```{r dispersion-estimate}
plotDispEsts(dds)
```

```{r deseq2-handler}
handle_deseq2 = function(dds, summarydata, column) {
  all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse="vs")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
```

```{r annotate}
annotate_df = function(df, dataset, filter, symbol) {
  require(biomaRt)
  ensembl = useMart('ensembl', dataset = dataset)
  annot.df = getBM(attributes=c(filter, symbol), filters=c(filter), values=rownames(df),
                   mart=ensembl)
  m = merge(df, annot.df, by.x="row.names", by.y=filter)
  colnames(m)[1] = "id"
  return(m)
}
```

## MA-plots
  
```{r DESeq-output, results='asis'}
all_results = handle_deseq2(dds, summarydata, condition)
for(i in seq(length(all_results))) {
  plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```


## Volcano-plot

```{r volcano, fig.height=8, fig.width=9}

# used for formatting labels on ggplots
fmt <- function(){ 
  function(x) format(x,nsmall = 1,scientific = FALSE)
}
volcano_density_plot <- function(stats, title="Volcano Plot with Marginal Distributions", pval.cutoff=0.05, lfc.cutoff=1, shade.colour="green", shade.alpha=0.25, point.colour="gray", point.alpha=0.75, point.outline.colour="darkgray", line.colour="gray") {
  # get range of log fold change and p-value values to setup plot borders
  range.lfc <- c(floor(min(stats$logFC)), ceiling(max(stats$logFC)))
  range.pval <- c(floor(min(-log10(stats$adj.P.Val))), ceiling(max(-log10(stats$adj.P.Val))))
  
  #make top plot - density plot with fold changes
  lfcd <- as.data.frame(cbind(density(stats$logFC)$x, density(stats$logFC)$y))
  hist_top <- ggplot(data=stats, aes(x=logFC))+
    geom_density(color=line.colour)+
    geom_ribbon(data=subset(lfcd, V1>lfc.cutoff),aes(x=V1,ymax=V2),ymin=0,fill=shade.colour, alpha=shade.alpha)+
    theme_bw()+ 
    theme(axis.title.x=element_blank())+
    theme(plot.margin=unit(c(3,-5.5,4,3), "mm") )+
    scale_x_continuous(limits = range.lfc, breaks = range.lfc[1]:range.lfc[2], expand = c(.05,.05))+
    scale_y_continuous(labels=fmt()) + labs(list(title="fold changes density plot"))
  
  # make blank plot
  empty <- ggplot()+geom_point(aes(1,1), colour="white")+
    theme(panel.grid=element_blank(),
          axis.ticks=element_blank(), 
          panel.background=element_blank(), 
          axis.text.x=element_blank(),  
          axis.text.y=element_blank(),          
          axis.title.x=element_blank(), 
          axis.title.y=element_blank()
          )
  
  #make scatter volcano plot
  scat.poly.up <- with(stats, data.frame(x=as.numeric(c(lfc.cutoff,  lfc.cutoff, max(range.lfc),max(range.lfc))), y=as.numeric(c(-log10(pval.cutoff), max(range.pval), max(range.pval),-log10(pval.cutoff)))))
  
  scatter <- ggplot(data=stats, aes(x=logFC, y=-log10(adj.P.Val))) +
    geom_point(alpha=point.alpha, pch=21, fill=point.colour, color=point.outline.colour) +
    geom_polygon(data=scat.poly.up, aes(x=x,y=y), fill=shade.colour, alpha=shade.alpha) +
    xlab("log2 fold change") + ylab("-log10(adjusted p-value)") +
    theme_bw()+
    theme(legend.position="none") +
    theme(plot.margin=unit(c(3,-5.5,4,3), "mm") )+
    scale_x_continuous(limits = range.lfc, breaks = range.lfc[1]:range.lfc[2], expand = c(.05,.05))+
    scale_y_continuous(labels=fmt(), limits = range.pval)+ labs(list(title="Volvano plot"))
  
  # make right plot - density plot of adjusted pvalues
  pvald <- as.data.frame(cbind(density(-log10(stats$adj.P.Val))$x, density(-log10(stats$adj.P.Val))$y))
  hist_right <- ggplot(data=stats, aes(x=-log10(adj.P.Val)))+
    geom_density(color=line.colour)+
    geom_ribbon(data=subset(pvald, V1>-log10(pval.cutoff)),aes(x=V1,ymax=V2),ymin=0,fill=shade.colour, alpha=shade.alpha)+
    theme_bw()+coord_flip()+
    scale_x_continuous(limits = range.pval)+
    theme(axis.title.y=element_blank())+ 
    theme(plot.margin=unit(c(3,-5.5,4,3), "mm")) + labs(list(title="adj.pval density plot"))
  
  # plot all plots
  pp.logfc <- ggplotGrob(hist_top)
  pp.empty <- ggplotGrob(empty)
  pp.volc <- ggplotGrob(scatter)
  pp.pval  <- ggplotGrob(hist_right)
  grid.arrange(main=title,arrangeGrob(pp.logfc,pp.volc, heights=c(1,3),ncol=1),
               arrangeGrob(pp.empty,pp.pval,  heights=c(1,3),ncol=1), 
               ncol=2, widths=c(3,1))
  } 

stats = as.data.frame(all_results[[1]])[,c(2,6)]
names(stats) = c("logFC","adj.P.Val")
stats = stats[!is.na(stats$adj.P.Val),]
volcano_density_plot(stats,title=names(all_results)[1],pval.cutoff=0.01, lfc.cutoff=1.5  )
```


## DEGreport
```{r get-groups}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp,"vs"))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[2]]
  list(g1,g2)
}
```

### Mean of DEG genes versus background

g1 and g2 correspond to both group, and r1 and r2 to random genes from both group.

```{r DEGreport-MB}
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], condition)
  detag <- row.names(subset(as.data.frame(all_results[[i]]), padj <= 0.1))
  print(degMB(detag, g[[1]], g[[2]], rlogMat) + 
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i])))
}
```

### Variance of DEG genes versus background

Same logic than before but with the variance of the group

```{r DEGreport-VB}
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], condition)
  detag <- row.names(subset(as.data.frame(all_results[[i]]), padj <= 0.1))
  print(degMB(detag, g[[1]], g[[2]], rlogMat) + 
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i])))
}
```

### Pvalues-vs-Mean

Pvalues are classified by the 10th quartiles of average gene expression.
The idea is to detect whether there is some bias to some group of genes due to
the expression.
```{r DEGreport-M}
for(i in seq(length(all_results))) {
  print(degMean(all_results[[i]]$pvalue, rlogMat) + 
  ggtitle(paste0("Pvalues-vs-Mean for ", names(all_results)[i])))
}
```

### Pvalues-vs-Variation
Pvalues are classified by the 10th quartiles of variance gene expression. 
```{r DEGreport-V}
for(i in seq(length(all_results))) {
  print(degVar(all_results[[i]]$pvalue, rlogMat) + 
  ggtitle(paste0("Pvalues-vs-Variation for ", names(all_results)[i])))
}
```

### Mean-vs-Variation
It shows the maximum variance of the two groups respect the minimum mean expression of the 
two groups. Significan genes and non-significan genes should follow the same 
correlation.
```{r DEGreport-MV}
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], condition)
  print(degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, rlogMat) + 
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i])))
}
```

## Differentially expressed genes

```{r DESeq-tables, results='asis'}
i =1
 write.table(counts,paste0(path_files,"counts.tsv"))

  out_df = annotate_df(data.frame(all_results[[i]]), dataset, filter, symbol)
  #out_df = all_results[[i]]
  write.table(out_df, file=paste0(path_files,names(all_results)[i], ".tsv"),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
 
knitr::kable(head(out_df[order(out_df$pvalue),],15))


```


[download results](`r get_report_links("~/repos/pipelines/glasgow/mmu_p53/Ctrlvsp53.tsv")`)

[download counts](`r get_report_links("~/repos/pipelines/glasgow/mmu_p53/counts.tsv")`)


### overlap with paper

```{r overlap, fig.width=11}
paper = read.table(paste0(path_files,"top40.dat"),header=T)
paper = melt(cbind(V1=1,paper),id.vars = "V1")
idx = match(paper$value,out_df$external_gene_name)
clean_idx = idx[!is.na(idx)]
dd = cbind(out_df[clean_idx,c(3,7,8)],paper=paper$variable[!is.na(idx)])
ggplot(dd, aes(x=external_gene_name,y=log2FoldChange,colour=cut(padj,breaks=c(-1,0.001,0.01,0.05,1.1)))) +
    geom_point()+
    scale_color_brewer("p.adj",palette = "Set1")+
    facet_wrap(~paper,nrow=2)+
    theme_bw()+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size = 10))
```


## heatmap with DE genes
```{r hetmap-de}
de = as.data.frame(all_results[[1]])
de = subset(de, padj<0.01)
detag = row.names(de)
pheatmap( rlogMat[detag,], show_rownames = F )
```

## Go ontology

```{r go-func}
get_entrez = function(g){
  require(biomaRt)
  ensembl = useMart('ensembl', dataset = dataset)
  df = getBM(attributes=c("ensembl_gene_id", "entrezgene"), filters="ensembl_gene_id", values=g,mart=ensembl)
  return(df[,2])
}

```


```{r do_go,cache=TRUE}
#library(org.Mm.eg.db)
#set <- universeGO(org.Mm.egGO,"Mus musculus")
#g_ezid = get_entrez(detag)
#g_ezid = g_ezid[!is.na(g_ezid)]
#mf <- runGO(g,set,"MF",as.list(GOMFCHILDREN))
library(gProfileR)
gp = gprofiler(detag, organism = "mmusculus")
write.table(gp[,1:13],paste0(path_files,"gprofile.tsv"))
```


[download go ontology](`r get_report_links("~/repos/pipelines/glasgow/mmu_p53/gprofile.tsv")`)


```{r table-go, results='asis',eval=TRUE}
knitr::kable(head(gp[order( gp$p.value ), c(12,2:11)],20))
```




