---
title: "RNA-seq analysis with k12 annotation."
output:
  html_document:
    toc: true
    toc_depth: 2
    theme: readable
    highlight: zenburn
---

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project=" bacterial - RNAseq"
labPI=" Souvik Ghosh "
lablocation="ManusBio"
analystname="Lorena Pantano"
analystemail="lpantano@hsph.harvard.edu"
```

RNA-seq analysis for`r labPI` at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 


```{r knitrsetup, echo=FALSE}
library(rmarkdown)
library(knitr)
knitr::opts_chunk$set(dev="png",echo=FALSE,
                      highlight=TRUE, autodep=TRUE,
                      comment='', fig.cap='',
                      tidy.opts=list(keep.blank.line=FALSE, width.cutoff=200),
                      cache=TRUE,warning=FALSE,message=FALSE,
                      error=FALSE, fig.width = 10, fig.height = 7)
options(bitmapType = 'cairo')
```

```{r project-options, echo=FALSE, cache=FALSE}
source("~/repos/myRfunctions/transactions.R")
path_results = "~/orch/scratch/Project_ManusBiosynthesis/rnaseq-k12/final/2016-02-04_rnaseq-k12/files"
dir.create(path_results, showWarnings = FALSE)
```

```{r create-report, echo=FALSE, eval=FALSE}
#render_2_drop("qc-summary-k12.Rmd", "manusbio_rnaseq_k12_Jan2016")
```

# Overview

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(DESeq2)
library(gridExtra)
library(org.EcK12.eg.db)
library(clusterProfiler)

project_summary = "/home/lpantano/orch/scratch/Project_ManusBiosynthesis/rnaseq-k12/final/2016-02-04_rnaseq-k12/project-summary.csv"
counts_file = "/home/lpantano/orch/scratch/Project_ManusBiosynthesis/rnaseq-k12/final/2016-02-04_rnaseq-k12/annotated_combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
symbol = as.data.frame(cbind(id=rownames(counts),symbol=as.character(counts[,"symbol"])))
counts = counts[,1:ncol(counts)-1]
counts = counts[, order(colnames(counts))]
colnames(counts) = gsub(".counts", "", colnames(counts))
counts[is.na(counts)] = 0
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads",
    "Median.insert.size", "Mapped.reads.pct", "Total.reads",
    "avg_coverage_per_region", "Mapped.Reads")
metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, clustering_distance_cols = "correlation", clustering_method = "ward.D2", ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Reads related metrics

### Mapped reads

```{r mapped-plot-noqualimap, eval="Mapped" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

### Genomic mapping rate
```{r mapping-rate-plot-noqualimap, eval="Mapping.rate" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

### rRNA and tRNA derived reads
```{r rRNA-sailfish}
ggplot(summarydata, aes(x=Name, y=rRNA_rate)) +
    geom_bar(stat="identity") +
    ylab("rRNA/tRNA rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

### Number of genes detected
```{r genes-detected-plot}
dd = data.frame(Name=names(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

### Gene detection saturation
```{r saturation-plot}
dd = data.frame(Mapped=summarydata$Mapped.reads, Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Mapped, y=Genes.Detected)) +
    geom_point() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped")
```

## Count related metrics

### Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

### Density of log10 TMM-normalized counts
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

### Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="spearman"))
```

### PCA plot
```{r pca}
dds = DESeqDataSetFromMatrix(countData=counts, colData=summarydata, design=~Name)
vst = varianceStabilizingTransformation(dds)
plotPCA(vst, intgroup=c("group"))
```

### Heatmap by concordance correlation coefficient
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075

```{r propcor-heatmap}
propcor = function(x, y) {
    x = log(x + 0.1)
    y = log(y + 0.1)
    num = 2 * cov(x, y)
    denom = var(x) + var(y)
return(num/denom)}

do_propcor = function(x) {
     mat = list()
     for(i in seq_len(ncol(x))) {
         for(j in seq_len(ncol(x))) {
        x2 = x[, i]
        y2 = x[, j]
        mat = c(mat, propcor(x2, y2)) } }
    mat = unlist(mat)
    mat = matrix(mat, ncol(x), ncol(x))
    colnames(mat) = colnames(x)
    rownames(mat) = colnames(x)
    return(mat)}

heatmap_fn(do_propcor(normalized_counts))
```
```{r de-setup}
library(DESeq2)
library(DEGreport)
library(vsn)
design = ~group
condition = "group"
```

# Differential expression

```{r deseq2-expression-analysis, results='asis'}
counts <- counts[rowSums(counts>0)>1,]
dds = DESeqDataSetFromMatrix(countData=counts,
    colData=summarydata, design = design)
dds = DESeq(dds)
```

## Effect of variance stabilization

```{r deseq-diagnostics, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

out_df = as.data.frame(rlogMat)
out_df$symbol = as.character(symbol$symbol[match(as.character(rownames(out_df)), symbol$id)])
write.table(out_df, file.path(path_results, "log2_normalized_counts.tsv"), quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)

out_df = as.data.frame(counts(dds, normalized=TRUE))
out_df$symbol = as.character(symbol$symbol[match(as.character(rownames(out_df)), symbol$id)])
write.table(out_df, file.path(path_results, "normalized_counts.tsv"), quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)


p1=meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1), plot=F)$gg + ggtitle("log2 transformation")
p2=meanSdPlot(rlogMat[notAllZero,], plot=F)$gg + ggtitle("variance stabilization transformation")
p3=meanSdPlot(vstMat[notAllZero,], plot=F)$gg + ggtitle("regularized log transformation")
grid.arrange(p1,p2,p3)
```

## Dispersion estimates

```{r dispersion-estimate}
plotDispEsts(dds)
```

```{r deseq2-handler}
handle_deseq2 = function(dds, summarydata, column, all_combs=NULL) {
  if (is.null(all_combs)){
    all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  }
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse="_vs_")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
```


```{r get-groups}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp,"_"))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[3]]
  list(g1,g2)
}
```

```{r DESeq-output, results='asis'}
all_combs = lapply(combn(levels(summarydata[,condition]), 2, simplify=FALSE)[1:9],
                   function(x){rev(x)})
all_combs[[10]] = c("C", "D")
all_results = handle_deseq2(dds, summarydata, condition, all_combs)
for(i in seq(length(all_results))) {
  cat(paste("## Lowest adjusted p-value hits for", names(all_results)[i]))
  out_df = as.data.frame(all_results[[i]])
  out_df$symbol = as.character(symbol$symbol[match(as.character(rownames(out_df)), symbol$id)])
  out_df$description = convertIDs(as.character(out_df$symbol), "SYMBOL", "GENENAME", org.EcK12.eg.db, "useFirst")
  
  cat("\n",paste(capture.output(summary(all_results[[i]]))[1:8], collapse = "<br>"),"\n")

  cat("\n\n### MA plot plot\n\n")
  DESeq2::plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
  
  cat("\n\n### Volcano plot\n\n")
  stats = as.data.frame(all_results[[i]][,c(2,6)])
  volcano_density_plot(stats, title=names(all_results)[i], lfc.cutoff=1.5)

  cat("\n\n### QC for DE genes mean/variance\n")
  g <- get_groups(summarydata, names(all_results)[i], condition)
  p = degMV(g[[1]], g[[2]], all_results[[i]]$padj, counts(dds,normalized=TRUE)) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
  print(p)
  
  cat("\n\n### Heatmap most significand, padj<0.05\n")
  sign = row.names(out_df)[out_df$padj<0.05 & !is.na(out_df$padj)]
  g = unlist(strsplit(names(all_results)[i], split = "_vs_"))
  samples = grepl(g[1], summarydata$group) | grepl(g[2], summarydata$group)
  heatmap_fn(vstMat[sign, samples], show_rownames = F)
  cat("\n")
  
  cat("\n\n### Top DE genes\n\n")
  print(knitr::kable(head(out_df)))
  fn = paste(names(all_results)[i], ".tsv", sep="")
  write.table(out_df, file.path(path_results, fn), quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n\nDifferential expression file at: ", fn)
  cat("\n\n")

  cat("\n\n### GO ontology of DE genes (logFC>3 and FDR < 1%):")
  .res = as.data.frame(out_df)
  .idx = .res$padj<0.01 & abs(.res$log2FoldChange) > 3
  .idx[is.na(.idx)] = FALSE
  .de = out_df$symbol[.idx]
  .accnum = convertIDs(.de, "SYMBOL", "ENTREZID", org.EcK12.eg.db, "useFirst")
  ego <- enrichGO(gene = .accnum[!is.na(.accnum)], 
                  organism = "ecolik12", ont = "BP", pAdjustMethod = "BH",
                  pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE)
  cat("\n\n")
  print(print_enrichGO(summary(ego), 30))
  cat("\n\n")
  fn = paste(names(all_results)[i], "_goenrichment.tsv", sep="")
  if (class(ego) == "enrichResult"){
    write.table(summary(ego), file.path(path_results, fn), quote=FALSE, sep="\t",
                row.names=TRUE, col.names=TRUE)
  }
  cat("\n\n### KEGG of DE genes (logFC>3 and FDR < 1%):")
  ego <- enrichKEGG(gene = .accnum[!is.na(.accnum)], use_internal_data = T,
                    organism = "ecolik12", readable = TRUE)
  cat("\n\n")
  print(print_enrichGO(summary(ego), 30))
  cat("\n\n")
  fn = paste(names(all_results)[i], "_KEGGenrichment.tsv", sep="")
  if (class(ego) == "enrichResult"){
    write.table(summary(ego), file.path(path_results, fn), quote=FALSE, sep="\t",
                row.names=TRUE, col.names=TRUE)
  }
  cat("\n\n")
}
```


#R Session Info

```{r compress, echo=FALSE, cache=FALSE}
name_res = compress_results(path_results, "rnaseq_k12")
```

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```

