---
title: "RNA-seq analysis."
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: hide
    theme: readable
    highlight: zenburn
---

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project=" human macrophages  - RNAseq "
labPI="Jeffrey"
lablocation="Longwood"
analystname="Lorena Pantano"
analystemail="lpantano@hsph.harvard.edu"
```

RNA-seq analysis for `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 


```{r knitrsetup, echo=FALSE}
library(rmarkdown)
library(knitr)
knitr::opts_chunk$set(dev="png",echo=FALSE,
                      highlight=TRUE, autodep=TRUE,
                      comment='', fig.cap='',
                      tidy.opts=list(keep.blank.line=FALSE, width.cutoff=200),
                      cache=TRUE,warning=FALSE,message=FALSE,
                      error=FALSE, fig.width = 12, fig.height = 9)
options(bitmapType = 'cairo')
```

```{r project-options, echo=FALSE}
library(myRfunctions)
path_results = "~/orch/scratch/jeffreys_rnaseq/hg_macro/final/files"
dir.create(path_results, showWarnings = FALSE)
```

```{r create-report, echo=FALSE, eval=FALSE}
# Rscript -e 'library(rmarkdown);library(myRfunctions);render_2_drop("mrna-summary.Rmd","jeffrey/hgmacro_model")'
```

# Overview

```{r qc-setup, results='asis'}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(DESeq2)
library(tximport)
library(logging)
library(gridExtra)
library(clusterProfiler)
library(org.Hs.eg.db)
library(cluster)
library(dplyr)

basicConfig()
root_path = "~/orch/scratch/jeffreys_rnaseq/hg_macro/final/2016-05-23_hg_macro"
project_summary = file.path(root_path, "project-summary.csv")
counts_file = file.path(root_path, "combined.counts")
tx2genes_file = file.path(root_path, "tx2gene.csv")

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")

summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]

if (file.exists(tx2genes_file)) {
  loginfo("Using gene counts calculated from the Sailfish transcript counts.")
  sf_files = file.path(root_path, "..",
                       rownames(summarydata), "sailfish",
                       "quant", "quant.sf")
  names(sf_files) = summarydata$Name
  tx2gene = read.table(tx2genes_file, sep=",")
  txi.salmon = tximport(sf_files, type="salmon", tx2gene = tx2gene,
                        reader=readr::read_tsv,
                        countsFromAbundance="lengthScaledTPM")
  counts = txi.salmon$counts
} else {
  loginfo("Using gene counts calculated from featureCounts.")
  counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
}

# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads",
    "Average.insert.size", "Mapped.reads.pct", "Total.reads",
    "avg_coverage_per_region", "Mapped.Reads", "description", "name2")
metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
rownames(metadata) = summarydata$Name

knitr::kable(metadata)

```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, clustering_method = "ward.D2", clustering_distance_cols = "correlation", ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Mapped reads
```{r mapped-plot, eval="Mapped" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("") 
```

## Genomic mapping rate
```{r mapping-rate-plot, eval="Mapping.Rate" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

```{r mapping-rate-plot-noqualimap, eval="Mapped.reads.pct" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name, y=Mapped.reads.pct)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Number of genes detected
```{r genes-detected-plot, eval="Genes.Detected" %in% colnames(summarydata)}
dd = data.frame(Name=colnames(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Gene detection saturation
```{r saturation-plot, eval="Genes.Detected" %in% colnames(summarydata)}
dd = data.frame(Mapped=summarydata$Mapped, Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Mapped, y=Genes.Detected)) +
    geom_point() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped")
```

## Exonic mapping rate
```{r exonic-mapping-plot, eval="Exonic.Rate" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

## rRNA|mitRNA mapping rate
```{r check-rRNA-eval}
eval_rRNA = "rRNA_rate" %in% colnames(summarydata) & !sum(is.na(summarydata$rRNA_rate)) == nrow(summarydata)
```
```{r rRNA-rate-plot, eval=eval_rRNA}
ggplot(summarydata, aes(x=Name, y=rRNA_rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(as.data.frame(counts))
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts, method="spearman"), clustering_distance_rows="correlation")
```

## MDS plot
```{r pca, eval=ncol(counts) > 1}
dds = DESeqDataSetFromTximport(txi.salmon, colData=metadata, design=~1)
vst = varianceStabilizingTransformation(dds)
# plotPCA(vst, intgroup=c("group")) + geom_text(aes(label=summarydata$Name),hjust = 0, nudge_x = 0.3)
mds(assay(vst), condition = metadata$group)
```

## Heatmap by concordance correlation coefficient
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075

```{r propcor-heatmap, eval=ncol(counts) > 1}
propcor = function(x, y) {
    x = log(x + 0.1)
    y = log(y + 0.1)
    num = 2 * cov(x, y)
    denom = var(x) + var(y)
return(num/denom)}

do_propcor = function(x) {
     mat = list()
     for(i in seq_len(ncol(x))) {
         for(j in seq_len(ncol(x))) {
        x2 = x[, i]
        y2 = x[, j]
        mat = c(mat, propcor(x2, y2)) } }
    mat = unlist(mat)
    mat = matrix(mat, ncol(x), ncol(x))
    colnames(mat) = colnames(x)
    rownames(mat) = colnames(x)
    return(mat)}

heatmap_fn(do_propcor(normalized_counts))
```


```{r de-setup}
library(DESeq2)
library(DEGreport)
library(vsn)
design = ~ group
condition = "group"
```

# Differential expression

```{r deseq2-expression-analysis, results='asis'}
metadata$comp = paste0(metadata$group, "_", metadata$treatment)
dds = DESeqDataSetFromTximport(txi.salmon,
    colData=metadata, design = ~ donor + comp)
dds2 = DESeq(dds, test="LRT", reduced = ~ 1)
rlogMat = assay(rlog(dds2))
```


## Dispersion estimates

```{r dispersion-estimate}
plotDispEsts(dds2)
```

```{r deseq2-handler}
handle_deseq2 = function(dds, summarydata, column, all_combs=NULL) {
  if (is.null(all_combs)){
    all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  }
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse="_vs_")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
```

```{r get-groups}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp,"_"))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[3]]
  list(g1,g2)
}
```

```{r DESeq-output, results='asis'}
combs = list(c("KOsp140_none", "control_none"), 
             c("KOsp140_LPS", "control_LPS"),
             c("control_none", "control_LPS"))
             #, c("KOsp140_none", "KOsp140_LPS"))
dds = DESeq(dds)
all_results = handle_deseq2(dds, colData(dds), "comp", combs)
deg_res = list()
for (name in names(all_results)){
    res = all_results[[name]]
    deg_res[[name]] = degResults(res=res, dds, rlogMat=rlogMat, name, 
                    FDR=0.05, FC=log2(1.5), group="treatment", xs="group",
                    org=org.Hs.eg.db, 
                    path_results = path_results)
}

```

```{r fpkm-2}
fpkm = read.table(file.path(root_path,"combined.fpkm"), header=T, row.names=1)
fpkm = cbind(id=rownames(fpkm), fpkm)
fpkm = annotate_df(fpkm ,"id", 'hsapiens_gene_ensembl', "ensembl_gene_id", "external_gene_name")
save_file(fpkm, "combined.fpkm", path_results)
rownames(fpkm) = fpkm$id
```

```{r fpkm}
fpkm = read.csv(file.path(root_path,"combined.gene.sf.tpm"),row.names = 1)
fpkm = cbind(id=rownames(fpkm), fpkm)
fpkm = annotate_df(fpkm ,"id", 'hsapiens_gene_ensembl', "ensembl_gene_id", "external_gene_name")
save_file(fpkm, "combined.tpm", path_results)
rownames(fpkm) = fpkm$id
```


```{r overlap-peaks}
peak0h = read.csv(file.path(root_path, "chipseq_peaks_0h.csv"), header = F)
peak4h = read.csv(file.path(root_path, "chipseq_peaks_4h.csv"), header = F)
peak0h_ens = convertIDs(as.character(peak0h$V1[peak0h$V1!="NONE"]), "SYMBOL", "ENSEMBL", org.Hs.eg.db)
peak4h_ens = convertIDs(as.character(peak4h$V1[peak4h$V1!="NONE"]), "SYMBOL", "ENSEMBL", org.Hs.eg.db)

save_file(fpkm[peak0h_ens[!is.na(peak0h_ens)],], "combined_peaks_0h_fpkm.csv", path_results)
save_file(fpkm[peak4h_ens[!is.na(peak4h_ens)],], "combined_peaks_4h_fpkm.csv", path_results)
```


# Custom figures

```{r custom-fn}
fmt <- function(){
    function(x) format(x,nsmall = 1,scientific = FALSE)
}
volcano_plot <- function(stats, side="both", title="Volcano Plot with Marginal Distributions",
                                 pval.cutoff=0.05, lfc.cutoff=1, shade.colour="green",
                                 shade.alpha=0.25, point.colour="gray", point.alpha=0.75,
                                 point.outline.colour="darkgray", line.colour="gray", plot_text=NULL) {
    require(grid)
    require(gridExtra)
    require(ggplot2)
    if ( !any(side %in% c("both","down","up")) | length(side)>1)
        stop("side parameter should be: both, up or down.")
    if (ncol(stats)!=2)
        stop("Need a data.frame with two columns: logFC and Adjusted.Pvalue")
    if ( sum( rowSums(is.na(stats)) ) > 0 )
        stats = stats[rowSums(is.na(stats))==0,]
    if ( any(stats[,2]>1) | any(stats[,2]<0) )
        stop("pvalues needs to be >0 and <1")
    names(stats) = c("logFC","adj.P.Val")
    stats[,2] = stats[,2] + 1e-10
    stats$color = point.colour
    stats$color[stats[,2]<pval.cutoff & (stats[,1]) > lfc.cutoff] = "red"
    stats$color[stats[,2]<pval.cutoff & (stats[,1]) < -lfc.cutoff] = "blue"
    # get range of log fold change and p-value values to setup plot borders
    range.lfc <- c(floor(min(stats$logFC)), ceiling(max(stats$logFC)))
    range.pval <- c(floor(min(-log10(stats$adj.P.Val))), ceiling(max(-log10(stats$adj.P.Val)))+1)

    #make top plot - density plot with fold changes
    # lfcd <- as.data.frame(cbind(density(stats$logFC)$x, density(stats$logFC)$y))
    # hist_top <- ggplot(data=stats, aes(x=logFC))+
    #     geom_density(color=line.colour)+
    #     theme_bw()+
    #     theme(axis.title.x=element_blank())+
    #     theme(plot.margin=unit(c(3,-5.5,4,3), "mm") )+
    #     scale_x_continuous(limits = range.lfc, breaks = range.lfc[1]:range.lfc[2], expand = c(.05,.05))+
    #     scale_y_continuous(labels=fmt()) + labs(list(title="fold changes density plot"))
    # if (side=="both" | side=="up")
    #     hist_top = hist_top + geom_ribbon(data=subset(lfcd, V1>lfc.cutoff),aes(x=V1,ymax=V2),ymin=0,fill=shade.colour, alpha=shade.alpha)
    # if (side=="both" | side=="down")
    #     hist_top = hist_top +  geom_ribbon(data=subset(lfcd, V1< -lfc.cutoff),aes(x=V1,ymax=V2),ymin=0,fill=shade.colour, alpha=shade.alpha)


    # make blank plot
    # empty <- ggplot()+geom_point(aes(1,1), colour="white")+
    #     theme(panel.grid=element_blank(),
    #           axis.ticks=element_blank(),
    #           panel.background=element_blank(),
    #           axis.text.x=element_blank(),
    #           axis.text.y=element_blank(),
    #           axis.title.x=element_blank(),
    #           axis.title.y=element_blank()
    #     )

    #make scatter volcano plot
    # scat.poly.up <- with(stats, data.frame(x=as.numeric(c(lfc.cutoff,  lfc.cutoff, max(range.lfc),max(range.lfc))), y=as.numeric(c(-log10(pval.cutoff), max(range.pval), max(range.pval),-log10(pval.cutoff)))))
    # scat.poly.down <- with(stats, data.frame(x=as.numeric(c(-lfc.cutoff,  -lfc.cutoff, min(range.lfc),min(range.lfc))), y=as.numeric(c(-log10(pval.cutoff), max(range.pval), max(range.pval),-log10(pval.cutoff)))))
    scatter <- ggplot(data=stats, aes(x=logFC, y=-log10(adj.P.Val))) +
        geom_point(alpha=point.alpha, pch=21, size=5, aes(fill=color), color=NA) +
        xlab("log2 fold change") + ylab("-log10(adjusted p-value)") +
        theme_bw()+
        theme(legend.position="none") +
        # theme(plot.margin=unit(c(3,-5.5,4,3), "mm") )+
        scale_color_manual(values = c("blue", point.colour, "red")) + 
        scale_fill_manual(values = c("blue", point.colour, "red")) + 
        scale_x_continuous(limits = range.lfc, breaks = range.lfc[1]:range.lfc[2], expand = c(.05,.05))+
        scale_y_continuous(labels=fmt(), limits = range.pval)+ labs(list(title="Volcano plot"))
    # if (side=="both" | side=="up")
    #     scatter = scatter + geom_polygon(data=scat.poly.up, aes(x=x,y=y), fill=shade.colour, alpha=shade.alpha)
    # if (side=="both" | side=="down")
    #     scatter = scatter + geom_polygon(data=scat.poly.down, aes(x=x,y=y), fill=shade.colour, alpha=shade.alpha)
    if (!is.null(plot_text)){
        require(ggrepel)
        names(plot_text) = c("logFC", "adj.P.Val", "name")
        plot_text[,2] <- plot_text[,2] + 1e-10
        scatter <- scatter + 
            # element_text(face="bold") + 
            geom_text_repel(data=plot_text, aes(x=logFC, y=-log10(adj.P.Val), label=name), size=4, fontface="bold")
    }
        
        
    # make right plot - density plot of adjusted pvalues
    # pvald <- as.data.frame(cbind(density(-log10(stats$adj.P.Val))$x, density(-log10(stats$adj.P.Val))$y))
    # hist_right <- ggplot(data=stats, aes(x=-log10(adj.P.Val)))+
    #     geom_density(color=line.colour)+
    #     geom_ribbon(data=subset(pvald, V1>-log10(pval.cutoff)),aes(x=V1,ymax=V2),ymin=0,fill=shade.colour, alpha=shade.alpha)+
    #     theme_bw()+coord_flip()+
    #     scale_x_continuous(limits = range.pval)+
    #     theme(axis.title.y=element_blank())+
    #     theme(plot.margin=unit(c(3,-5.5,4,3), "mm")) + labs(list(title="adj.pval density plot"))

    # plot all plots
    # pp.logfc <- ggplotGrob(hist_top)
    # pp.empty <- ggplotGrob(empty)
    pp.volc <- ggplotGrob(scatter)
    # pp.pval  <- ggplotGrob(hist_right)
    # p = grid.arrange(top=textGrob(title),
    #              arrangeGrob(pp.logfc,pp.volc, heights=c(1,3),ncol=1),
    #              arrangeGrob(pp.empty,pp.pval,  heights=c(1,3),ncol=1),
    #              ncol=2, widths=c(3,1))
    scatter
}
```


```{r load-data}
library(readr)
fns = unlist(sapply(list.files(path_results, pattern = ".tsv", full.names = T),
             function(x){
                 if (!grepl("goenrich", x) & !grepl("counts",x)){
                     return(x)
                 }
}))

fns=fns[!is.null(fns)]

res = list()
for (fn in fns){
    name = gsub(".tsv", "", basename(fn))
    res[[name]] = read_csv(fn)    
}

rlogMat = as.matrix(read.csv(file.path(path_results,"KOsp140_none_vs_control_none_log2_counts.csv"), row.names = 1, check.names = F))

```

## Heatmaps of DE genes

```{r heatmaps}

library(RColorBrewer)
my_palette <- rev(brewer.pal(11,"RdBu"))
col_order = c(c(1,5,3,7), c(2,6,4,8))

.get_text = function(data_text){
    data_text = annotate_df(data_text ,"id", 'hsapiens_gene_ensembl', "ensembl_gene_id", "external_gene_name")
    data_text[is.na(data_text$external_gene_name),"external_gene_name"] <- data_text[is.na(data_text$external_gene_name),"id"]
    as.data.frame(data_text[,c(2,3,4)])
}


for (name in names(res)[2:3]){
    df = as.data.frame(res[[name]])
    row.names(df) = df$id
    diff_genes_top = abs(df$log2FoldChange)>log2(1.5) & df$padj<0.05 & !is.na(df$padj)
    diff_genes_top = df$id[diff_genes_top]
    # diff_genes_top = diff_genes_top[1:10]
    hr = hclust(as.dist(1-cor(t(rlogMat[diff_genes_top,col_order]))), method = "ward.D2")
    pheatmap(rlogMat[hr$labels[hr$order],col_order],
              annotation=metadata,
              clustering_method = "ward.D2", 
              clustering_distance_cols = "correlation",
              scale='row', labels_row = "",
              cluster_row=FALSE, cluster_cols=FALSE,
              show_rownames=T, color=my_palette,main = name)
    fn_out = file.path(paste0(name, "_order_heatmap.csv"))
    out_sort = cbind(data.frame(ensembl=hr$labels[hr$order], 
                          symbol=convertIDs(hr$labels[hr$order], "ENSEMBL", "SYMBOL", org.Hs.eg.db)), df[hr$labels[hr$order],])
    save_file(out_sort, fn_out, path_results)
    df = df[!is.na(df$symbol),]
    stats = as.data.frame(df[!is.na(df$symbol),c(3,7)])
    text = read.csv(file.path(root_path, paste0(name, "_selected.csv")))
    text = rbind(df[as.character(text$id),c(3,7,8)], 
           (df[order(df$pvalue) ,c(3,7,8)])[1:5,],
           (df[order(df$log2FoldChange, decreasing = T) ,c(3,7,8)])[1:5,],
           (df[order(df$log2FoldChange) ,c(3,7,8)])[1:5,])
    p = volcano_plot(stats, title=name, lfc.cutoff=log2(1.5), pval.cutoff = 0.05, point.alpha=0.50, plot_text = text)
    print(p)
    p = volcano_plot(stats, title=name, lfc.cutoff=log2(1.5), pval.cutoff = 0.05, point.alpha=0.50)
    print(p)
}

```

## Heatmap at KO_LPS vs control_LPS only at 4h time

```{r custom-heatmap-condition}
name = "KOsp140_LPS_vs_control_LPS"
df = as.data.frame(res[[name]])
row.names(df) = df$id
diff_genes_top = abs(df$log2FoldChange)>log2(1.5) & df$padj<0.05 & !is.na(df$padj)
diff_genes_top = df$id[diff_genes_top]
col_keep = c(2,6,4,8)
hr = hclust(as.dist(1-cor(t(rlogMat[diff_genes_top,col_keep]))), method = "ward.D2")
pheatmap(rlogMat[hr$labels[hr$order],col_keep],
         annotation=metadata,
         clustering_method = "ward.D2", 
         clustering_distance_cols = "correlation",
         scale='row', labels_row = "",
         cluster_row=FALSE, cluster_cols=FALSE,
         show_rownames=T, color=my_palette,main = name)
fn_out = file.path(paste0(name, "_4honly_order_heatmap.csv"))
out_sort = cbind(data.frame(ensembl=hr$labels[hr$order], 
                            symbol=convertIDs(hr$labels[hr$order], "ENSEMBL", "SYMBOL",
                                              org.Hs.eg.db)),df[hr$labels[hr$order],])
save_file(out_sort, fn_out, path_results)

diff_genes_top = abs(df$log2FoldChange)>log2(1.5) & !is.na(df$padj)
diff_genes_top = df$id[diff_genes_top]
col_keep = c(2,6,4,8)
hr = hclust(as.dist(1-cor(t(rlogMat[diff_genes_top,col_keep]))), method = "ward.D2")
pheatmap(rlogMat[hr$labels[hr$order],col_keep],
         annotation=metadata,
         clustering_method = "ward.D2", 
         clustering_distance_cols = "correlation",
         scale='row', labels_row = "",
         cluster_row=FALSE, cluster_cols=FALSE,
         show_rownames=T, color=my_palette,main = paste0(name," only FC(1.5) cutoff"))
fn_out = file.path(paste0(name, "_4honly_fconly_order_heatmap.csv"))
out_sort = cbind(data.frame(ensembl=hr$labels[hr$order], 
                            symbol=convertIDs(hr$labels[hr$order], "ENSEMBL", "SYMBOL",
                                              org.Hs.eg.db)),df[hr$labels[hr$order],])
save_file(out_sort, fn_out, path_results)

```


## Genes up-regulated at Control_LPS and DE at KO_LPS vs control_LPS
```{r custom-de}
name = "KOsp140_LPS_vs_control_LPS_filtered_by_LPS_induced"
df = as.data.frame(res[[2]])
row.names(df) = df$id
diff_genes_top = abs(df$log2FoldChange)>log2(1.5) & df$padj<0.05 & !is.na(df$padj)
diff_genes_top = df$id[diff_genes_top]

lps = as.data.frame(res[[1]])
row.names(lps) = lps$id
diff_lps = lps$log2FoldChange< -log2(1.5) & lps$padj<0.05 & !is.na(lps$padj)
diff_lps = lps$id[diff_lps]

keep = intersect(diff_genes_top, diff_lps)
hr = hclust(as.dist(1-cor(t(rlogMat[keep,col_order]))), method = "ward.D2")
pheatmap(rlogMat[hr$labels[hr$order],col_order],
         annotation=metadata,
         clustering_method = "ward.D2", 
         clustering_distance_cols = "correlation",
         scale='row', labels_row = "",
         cluster_row=FALSE, cluster_cols=FALSE,
         show_rownames=T, color=my_palette,main = name)
fn_out = file.path(paste0(name, "_order_heatmap.csv"))
out_sort = cbind(data.frame(ensembl=hr$labels[hr$order], 
                            symbol=convertIDs(hr$labels[hr$order], "ENSEMBL", "SYMBOL", org.Hs.eg.db),
                            log2FC_LPS = -1 * lps[hr$labels[hr$order], "log2FoldChange"]),
                 df[hr$labels[hr$order],])
save_file(out_sort, fn_out, path_results)

ma = rlogMat[keep,c(2,6,4,8)]
hr = hclust(as.dist(1-cor(t(ma))), method = "ward.D2")
pheatmap((ma[hr$labels[hr$order],]),
         clustering_method = "ward.D2", 
         clustering_distance_cols = "correlation",
         scale='row', labels_row = "",
         cluster_row=FALSE, cluster_cols=FALSE,
         show_rownames=T, color=my_palette,main = name)
fn_out = file.path(paste0(name, "_4h_order_heatmap.csv"))
out_sort = cbind(data.frame(ensembl=hr$labels[hr$order], 
                            symbol=convertIDs(hr$labels[hr$order], "ENSEMBL", "SYMBOL", org.Hs.eg.db),
                            log2FC_LPS = -1 * lps[hr$labels[hr$order], "log2FoldChange"]),
                 df[hr$labels[hr$order],])
save_file(out_sort, fn_out, path_results)


ma = rlogMat[keep,col_order]
ma = data.frame(row.names=rownames(ma), Hi4h=rowMeans(ma[,5:6])-rowMeans(ma[,1:2]), Si4h=rowMeans(ma[,7:8])-rowMeans(ma[,3:4]))

hr = hclust(dist(abs(ma), method="minkowski"), method = "ward.D2")
pheatmap((ma[hr$labels[hr$order],]),
         clustering_method = "ward.D2", 
         clustering_distance_cols = "correlation",
         scale='none', labels_row = "",
         cluster_row=FALSE, cluster_cols=FALSE,
         show_rownames=T, color=my_palette,main = paste(name, "(log2FC)"))
fn_out = file.path(paste0(name, "_4h_mean_order_heatmap.csv"))
out_sort = cbind(data.frame(ensembl=hr$labels[hr$order], 
                            symbol=convertIDs(hr$labels[hr$order], "ENSEMBL", "SYMBOL", org.Hs.eg.db),
                            log2FC_LPS = -1 * lps[hr$labels[hr$order], "log2FoldChange"]),
                 df[hr$labels[hr$order],])
save_file(out_sort, fn_out, path_results)

```

## Heatmap of mouse data
```{r custom-heatmap-mouse}
df = read.table("~/orch/scratch/jeffreys_rnaseq/mouse_gene_list.txt", 
                sep="\t", comment.char = "#", header=T)
col_keep = c(6:11)
hr = hclust(as.dist(1-cor(t(df[,col_keep]))), method = "ward.D2")
pheatmap(df[hr$order,col_keep],
         clustering_method = "ward.D2", 
         clustering_distance_cols = "correlation",
         scale='row', labels_row = "",
         cluster_row=FALSE, cluster_cols=FALSE,
         show_rownames=T, color=my_palette,main = "mouse data")
fn_out = file.path(paste0("mouse_data.csv"))
out_sort = data.frame(gene=df$Symbol[hr$order])
save_file(out_sort, fn_out, path_results)
```

## Heatmaps of SP140 peaks
```{r heatmap-sp140}
.make_heatmap = function(ma, name, row=F, scale="row"){
    hr = hclust(as.dist(1-cor(t(ma))), method = "ward.D2")
     pheatmap(ma[hr$labels[hr$order],],
              clustering_method = "ward.D2", 
              clustering_distance_cols = "correlation",
              scale=scale, 
              cluster_row=FALSE, cluster_cols=FALSE,
              show_rownames=row, color=my_palette,main = name)
    fn_out = file.path(path_results, paste0(name, "_order_heatmap.csv"))
    out_sort = data.frame(symbol=hr$labels[hr$order])
    write.table(out_sort, fn_out, row.names=F, sep="\t")

}

sp140_10pct = read.csv(file.path(root_path, "SP140.10pct.peaksTPM.csv")) %>% distinct()
name = "SP140_10pct"
sp140_10pct = sp140_10pct[rowSums(sp140_10pct[2:5])>0, ]
rownames(sp140_10pct) = sp140_10pct[,1]
.make_heatmap(sp140_10pct[,2:5], name)


sp140_100 = read.csv(file.path(root_path, "SP140.100.peaksTPM.csv")) %>% distinct()
name = "SP140_100"
sp140_100 = sp140_100[rowSums(sp140_100[2:5])>0, ]
rownames(sp140_100) = sp140_100[,1]
.make_heatmap(sp140_100[,2:5], name)

```

## HOX genes
```{r heatmap-hox}
.summarize = function(ma, groups){
    sapply(levels(groups), function(g){
        rowMeans(ma[,g==groups])
    })
}
groups = factor(gsub("^.[.]", "", colnames(fpkm)[2:9]), levels=c("Hi0h",
                                                          "Hi4h",
                                                          "Si0h",
                                                          "Si4h"))

hox = read.csv(file.path(root_path, "hox_genes.csv"), header=F) %>% distinct()
idx = match(hox$V1, fpkm$external_gene_name)
idx = idx[!is.na(idx)]
ma = fpkm[idx,2:9]
ma = ma[rowSums(ma>0)>1,]
ma = .summarize(ma, groups)
rownames(ma) = fpkm[rownames(ma),"external_gene_name"]

.make_heatmap(ma[,c(1,3,2,4)], "HOX", row=T)

.make_heatmap(ma[,c(1,3,2,4)], "HOX", row=T, scale="none")
```

## HOX genes - list 2
```{r heatmap-hox-1-13}
hox = read.csv(file.path(root_path, "hox_genes_list2.csv"), header=F) %>% distinct()
idx = match(hox$V1, fpkm$external_gene_name)
idx = idx[!is.na(idx)]
ma = fpkm[idx,2:9]
ma = ma[rowSums(ma>0)>0,]
ma = .summarize(ma, groups)
rownames(ma) = fpkm[rownames(ma),"external_gene_name"]
.make_heatmap(ma[,c(1,3,2,4)], "HOX", row=T)

.make_heatmap(ma[,c(1,3,2,4)], "HOX", row=T, scale="none")
```

```{r fpkm-iso, echo=F, eval=F}
fpkm = read.table(file.path(root_path,"combined.isoform.sf.tpm"),row.names = 1,header=T)
fpkm = cbind(id=rownames(fpkm), fpkm)
fpkm = annotate_df(fpkm ,"id", 'hsapiens_gene_ensembl', "ensembl_transcript_id", "external_gene_name")
save_file(fpkm, "combined.tpm", path_results)
rownames(fpkm) = fpkm$id
```


#R Session Info

```{r compress, echo=FALSE, cache=FALSE}
name_res = compress_results(path_results, prefix = "hgmacro_results")
```

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```
