---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

Last updated `r date()` by Lorena Pantano

```{r setup, echo=FALSE}
options(bitmapType = "cairo")
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", bootstrap.show.code = FALSE, fig.width = 9,
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

```{r render}
library(rmarkdown)
library(knitrBootstrap)
source("~/repos/myRfunctions/transactions.R")
# render_2_drop("qc-summary.Rmd", dn = "jeffrey_chron_w_snps")
```

# Overview

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(gridExtra)
library(dplyr)
library(knitr)
project_summary = "/home/lpantano/orch/scratch/jeffreys_chron_snp/chron_snp/final/2015-10-01_chron_snp/project-summary.csv"
counts_file = "/home/lpantano/orch/scratch/jeffreys_chron_snp/chron_snp/final/2015-10-01_chron_snp/combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]
colnames(counts) = gsub(".counts", "", colnames(counts))
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA.rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias")
```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, clustering_distance_rows = "correlation", clustering_method = "ward", ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Mapped reads
```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

## Genomic mapping rate
```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Number of genes detected
```{r genes-detected-plot}
dd = data.frame(Name=names(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Gene detection saturation
```{r saturation-plot}
dd = data.frame(Mapped=summarydata$Mapped, Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Mapped, y=Genes.Detected)) +
    geom_point() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped")
```

## Exonic mapping rate
```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation (Pearson) heatmap of TMM-normalized counts
```{r pearson-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="pearson"))
```

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="spearman"))
```

## MDS plot of TMM-normalized counts
```{r mds-normalized}
mds(normalized_counts, k=length(colnames(normalized_counts)) - 1, d = "cor")
```

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select = order(rowMeans(normalized_counts),decreasing=TRUE)[1:30]
heatmap_fn(log2(normalized_counts[select,]+1))
```

## Heatmap by concordance correlation coefficient
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075

```{r propcor-heatmap}
propcor = function(x, y) {
    x = log(x + 0.1)
    y = log(y + 0.1)
    num = 2 * cov(x, y)
    denom = var(x) + var(y)
return(num/denom)}

do_propcor = function(x) {
     mat = list()
     for(i in seq_len(ncol(x))) {
         for(j in seq_len(ncol(x))) {
        x2 = x[, i]
        y2 = x[, j]
        mat = c(mat, propcor(x2, y2)) } }
    mat = unlist(mat)
    mat = matrix(mat, ncol(x), ncol(x))
    colnames(mat) = colnames(x)
    rownames(mat) = colnames(x)
    return(mat)}

heatmap_fn(do_propcor(normalized_counts))
```

# Differential expression

```{r de-setup}
library(DESeq2)
library(DEGreport)
library(vsn)
```

```{r time-deseq}
summarydata$covariate = paste0(summarydata$time, summarydata$condition, summarydata$genotype)

```

```{r time-deseq2-expression-analysis, results='asis', cache=T}
counts <- counts[rowSums(counts>1)>2,]
dds = DESeqDataSetFromMatrix(countData=counts,
    colData=summarydata, design = ~ covariate)
dds = DESeq(dds)
rld = rlog(dds)
```

## Dispersion estimates

```{r time-dispersion-estimate}
plotDispEsts(dds)
```

## PCAplot of 50 most variant genes

```{r time-pca}
plotPCA(rld, "covariate", n=100)
```

```{r time-deseq2-handler}
handle_deseq2 = function(dds, summarydata, column, all_combs) {
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse="_vs_")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
.make_binary_matrix = function(ma){
  total_genes = as.data.frame(matrix(nrow=length(unique(unlist(ma))),ncol=length(ma)))
  total_genes[] = 0
  rownames(total_genes) = unique(unlist(ma))
  colnames(total_genes) = names(ma)
  for (comp in names(ma)){
    total_genes[ma[[comp]],comp] = 1
  }
  total_genes$Name = as.factor(rownames(total_genes))
  rownames(total_genes) = 1:nrow(total_genes)
  total_genes = total_genes[,c(4,1,2,3)]
  for (nc in 2:4)
    total_genes[,nc] <- as.integer(total_genes[,nc])
  total_genes  
}
library(clusterProfiler)

reduce = function(genes){
  seen = c()
  idx = sapply(genes, function(x){
    here = as.character(as.vector(unlist(strsplit(x, split = "/"))))
    if (length(here)>100){
      return(FALSE)
    }
    c = intersect(seen, here)
    seen <<- unique(c(seen, here))
    score = 0.6 * length(here)
    if (length(c) < score)
      return(TRUE)
    FALSE
  })
  idx
}

goenrichment = function(g){
  select_entrez = .ens2entrez(g)
  ego= enrichGO(gene = select_entrez, organism = "human", ont = "BP", pAdjustMethod = "BH",
  pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE)

  if (nrow(summary(ego))>0){
    idx =reduce(summary(ego)$geneID)
    
    cat("GO enrichment for Biological processes:\n\n")
    
    print(kable(summary(ego)[idx, 1:7]))
    
    cat("\n\n")
  }  
}
  
library(org.Hs.eg.db)
.ens2name = function(x){
    symbol = AnnotationDbi::select(org.Hs.eg.db, as.character(x), "SYMBOL", keytype="ENSEMBL")
    symbol = symbol %>% distinct(SYMBOL)
    idx = match(x, symbol$ENSEMBL)
    # symbol$ENTREZID[!is.na(symbol$ENTREZID)]
    symbol$SYMBOL[idx]
}

.ens2entrez = function(x){
    symbol = AnnotationDbi::select(org.Hs.eg.db, as.character(x), "ENTREZID", keytype="ENSEMBL")
    symbol = symbol %>% distinct(ENTREZID)
    symbol$ENTREZID
}

```

## Differentially expressed genes between 4h and 0h

comparisons:

* Wt 4h vs 0h = 4hwtnon vs 0hwtnon
* Cd with snp 4h vs 0h = 4hcdsnp vs 0hcdsnp
* Cd without snp 4h vs 0h = 4hcdnon vs 0hcdnon

### MA-plots

```{r time-DESeq-output, results='asis'}
comps = list(wt=c("4hwtnon", "0hwtnon"),
             cdnon=c("4hcdnon", "0hcdnon"),
             cdsnp=c("4hcdsnp", "0hcdsnp"))
all_results = handle_deseq2(dds, summarydata, "covariate", comps)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
scale_factor=9
for(i in seq(length(all_results))) {
  DESeq2::plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

```{r time-get-groups}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp,"_vs_"))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[2]]
  list(g1,g2)
}
```

### Mean-vs-Variation

```{r time-DEGreport-MV}
plots = list()
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], "covariate")
  plots[[i]] = degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, counts(dds,normalized=TRUE)) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Differentially expressed genes tables

FDR < 1% and abs(log2FC) > 2

```{r time-DESeq-tables, results='asis', cache=T}
ma = list()
for(i in seq(length(all_results))) {
  cat(paste("\n\nLowest adjusted p-value hits for", names(all_results)[i]),"\n\n")
  out_df = all_results[[i]]
  print(summary(out_df,alpha=0.01))
  cat("\n\n")
  out_df$name = .ens2name(rownames(out_df))
  genes = as.data.frame(out_df) %>% mutate(ensembl=rownames(out_df)) %>% filter(abs(log2FoldChange)>2 & padj<0.01)
  ma[[paste0("c",names(all_results)[i])]] = genes$ensembl
  
  print(knitr::kable(head(out_df,20)))
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n")
  
  goenrichment(genes$ensembl)
  
}
```

### Common genes between comparisons

```{r time-upset}
library(UpSetR)
de_genes = c(unlist(ma))
binary_ma=.make_binary_matrix(ma)
upset(binary_ma, sets = names(ma))
venn(ma)
corder = colnames(counts)[c(seq(1,12,2),seq(2,12,2))]
corder = corder[c(5:6,1:4,11:12,7:10)]
library(RColorBrewer)
my_palette <- rev(brewer.pal(11,"RdBu"))
heatmap_fn(as.matrix(assay(rld)[de_genes,corder]), scale='row',show_rownames=F,cluster_cols=FALSE, color=my_palette)
```

## Differentially expressed genes between groups at 0h

comparisons:

* Wt vs Cd with snp at 0h = 0hwtnon vs 0hcdsnp
* Wt vs Cd without snp at 0h = 0hwtnon vs 0hcdnon
* Cd with snp vs Cd without snp at 0h = 0hcdsnp vs 0hcdnon

### MA-plots

```{r group-DESeq-output, results='asis'}
comps = list(wt=c("0hwtnon", "0hcdnon"),
             cdnon=c("0hwtnon", "0hcdsnp"),
             cdsnp=c("0hcdsnp", "0hcdnon"))
all_results = handle_deseq2(dds, summarydata, "covariate", comps)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  DESeq2::plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

### Mean-vs-Variation

```{r group-DEGreport-MV}
plots = list()
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], "covariate")
  plots[[i]] = degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, counts(dds,normalized=TRUE)) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Differentially expressed genes

FDR < 1% and abs(log2FC) > 1

```{r group-DESeq-tables, results='asis', cache=T}
ma = list()
for(i in seq(length(all_results))) {
  cat(paste("\n\nLowest adjusted p-value hits for", names(all_results)[i]), "\n\n")
  out_df = all_results[[i]]
  print(summary(out_df,alpha=0.01))
  cat("\n\n")
  out_df$name = .ens2name(rownames(out_df))
  genes = as.data.frame(out_df) %>% mutate(ensembl=rownames(out_df)) %>% filter(abs(log2FoldChange)>1 & padj<0.01)
  ma[[paste0("c",names(all_results)[i])]] = genes$ensembl
  
  print(knitr::kable(head(out_df,20)))
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n")
  
  goenrichment(genes$ensembl)
  
  
}
```

### Common genes between comparisons

```{r group-upset}
de_genes = c(de_genes, unlist(ma))
binary_ma=.make_binary_matrix(ma)
upset(binary_ma, sets = names(ma))
venn(ma)
heatmap_fn(as.matrix(assay(rld)[de_genes,corder]), scale='row',show_rownames=F,cluster_cols=FALSE, color=my_palette)
```

## Differentially expressed genes between groups at 4h

comparisons:

* Wt vs Cd with snp at 4h = 4hwtnon vs 4hcdsnp
* Wt vs Cd without snp at 4h = 4hwtnon vs 4hcdnon
* Cd with snp vs Cd without snp at 4h = 4hcdsnp vs 4hcdnon

### MA-plots

```{r group4h-DESeq-output, results='asis'}
comps = list(wt=c("4hwtnon", "4hcdnon"),
             cdnon=c("4hwtnon", "4hcdsnp"),
             cdsnp=c("4hcdsnp", "4hcdnon"))
all_results = handle_deseq2(dds, summarydata, "covariate", comps)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  DESeq2::plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

### Mean-vs-Variation

```{r group4h-DEGreport-MV}
plots = list()
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], "covariate")
  plots[[i]] = degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, counts(dds,normalized=TRUE)) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Differentially expressed genes

FDR < 1% and abs(log2FC) > 1

```{r group4h-DESeq-tables, results='asis', cache=T}
ma = list()
for(i in seq(length(all_results))) {
  cat(paste("\n\nLowest adjusted p-value hits for", names(all_results)[i]),"\n\n")
  out_df = all_results[[i]]
  print(summary(out_df,alpha=0.01))
  cat("\n\n")
  genes = as.data.frame(out_df) %>% mutate(ensembl=rownames(out_df)) %>% filter(abs(log2FoldChange)>1 & padj<0.01)
  ma[[paste0("c",names(all_results)[i])]] = genes$ensembl
  out_df$name = .ens2name(rownames(out_df))
  
  print(knitr::kable(head(out_df,20)))
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n")
  
  goenrichment(genes$ensembl)
  
}
```

### Common genes between comparisons

```{r group4h-upset}
de_genes = c(de_genes, unlist(ma))
binary_ma=.make_binary_matrix(ma)
upset(binary_ma, sets = names(ma))
venn(ma)
heatmap_fn(as.matrix(assay(rld)[de_genes,corder]), scale='row',show_rownames=F,cluster_cols=FALSE, color=my_palette)

```

## Heatmap with all DE genes

```{r heatmpa-all}
de_genes = unlist(unique(de_genes))
ma = as.matrix(assay(rld)[de_genes,corder])
hr = hclust(as.dist(1-cor(t(ma))), method = "ward.D2")
heatmap_fn(ma[hr$order,], scale='row', cluster_row=FALSE, cluster_cols=FALSE, show_rownames=F, color=my_palette)
name_genes = .ens2name(hr$labels[hr$order])
name_genes[is.na(name_genes)] = hr$labels[hr$order][is.na(name_genes)]
write.table(data.frame(symbol=name_genes, ensembl=hr$labels[hr$order]),"all_de_genes_order.tsv",row.names = F,col.names = F)

heatmap_fn(ma[hr$order,], scale='row', cluster_row=FALSE, show_rownames=F, color=my_palette)
```

[genes sorted](`r get_report_links("all_de_genes_order.tsv")`)

# Custom hetmpas

sheet 1 in excel

```{r sheet-1}
genes = read.table("unique_hc_v_snp_plus_0_4.csv", skip=1, stringsAsFactors = FALSE)
ma = as.matrix(assay(rld)[,corder])
name_genes = .ens2name(row.names(ma))
rownames(ma)[!is.na(name_genes)] = name_genes[!is.na(name_genes)]
genes = intersect(genes$V1, rownames(ma))

ma_sub = ma[genes,]
hr = hclust(as.dist(1-cor(t(ma_sub))), method = "ward.D2")
heatmap_fn(ma_sub[hr$order,], scale='row', cluster_row=FALSE, cluster_cols=FALSE, show_rownames=F, color=my_palette)
name_genes = hr$labels[hr$order]
write.table(data.frame(symbol=name_genes),"sheet1_genes_order.tsv",row.names = F,col.names = F)
```

sheet 2 in excel

```{r sheet-2}
genes = read.table("unique_hc_v_snpplus_minus_hc_snpminus.csv", skip=1, stringsAsFactors = FALSE)
genes = intersect(genes$V1, rownames(ma))
ma_sub = ma[genes,]
hr = hclust(as.dist(1-cor(t(ma_sub))), method = "ward.D2")
heatmap_fn(ma_sub[hr$order,], scale='row', cluster_row=FALSE, cluster_cols=FALSE, show_rownames=F, color=my_palette)
name_genes = hr$labels[hr$order]
write.table(data.frame(symbol=name_genes),"sheet2_genes_order.tsv",row.names = F,col.names = F)
```

sheet 3 in excel

```{r sheet-3}
genes = read.table("lps_iduced_snppllus_snpminus.csv", skip=1, stringsAsFactors = FALSE)
genes = intersect(genes$V1, rownames(ma))
ma_sub = ma[genes,]
hr = hclust(as.dist(1-cor(t(ma_sub))), method = "ward.D2")
heatmap_fn(ma_sub[hr$order,], scale='row', cluster_row=FALSE, cluster_cols=FALSE, show_rownames=F, color=my_palette)
name_genes = hr$labels[hr$order]
write.table(data.frame(symbol=name_genes),"sheet3_genes_order.tsv",row.names = F,col.names = F)
```

# Files

```{r files, results='asis'}
fn = list.files(".", pattern="tsv")
for (f in fn){
  link=get_report_links(f)
  cat(sep="", "[",f,"]","(",link,")\n\n")
}
```

```{r fpkm}
fpkm = read.table("combined.fpkm", header=T, row.names=1)
fpkm$name = .ens2name(rownames(fpkm))
fpkm$ens = rownames(fpkm)
write.table(fpkm[,c(13,14,1:12)], "combined_w_name.fpkm", quote=F,sep="\t", row.names=F)
```

[fpkm](`r get_report_links("combined_w_name.fpkm")`)

# Isoforms

```{r iso-fpkm, results='asis'}
iso_fpkm = read.table("~/orch/scratch/jeffreys_chron_snp/chron_snp/work/htseq-count/combined.isoform.fpkm", header=T)
iso_names = read.table("~/orch/scratch/jeffreys_chron_snp/chron_snp/work/htseq-count/isoform_to_gene.txt")
idx = match(iso_fpkm$id, iso_names$V1)
names = .ens2name(iso_names$V2[idx])
iso_fpkm$name = names
write.table(iso_fpkm, "combined_iso.fpkm", row.names=F, quote=F, sep="\t")

kable(rbind(iso_fpkm[grepl("SP140",names),],
            iso_fpkm[grepl("SP110",names),],
            iso_fpkm[grepl("SP100",names),]))

```

[fpkm](`r get_report_links("combined_iso.fpkm")`)
