---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

Last updated `r date()`

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
project_summary = "/home/lpantano/orch/scratch/jeffreys_chron_snp/chron_snp/final/2015-10-01_chron_snp/project-summary.csv"
counts_file = "/home/lpantano/orch/scratch/jeffreys_chron_snp/chron_snp/final/2015-10-01_chron_snp/combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]
colnames(counts) = gsub(".counts", "", colnames(counts))
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA.rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias")
```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, clustering_distance_rows = "correlation", clustering_method = "ward",...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Mapped reads
```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

## Genomic mapping rate
```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Number of genes detected
```{r genes-detected-plot}
dd = data.frame(Name=names(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Gene detection saturation
```{r saturation-plot}
dd = data.frame(Mapped=summarydata$Mapped, Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Mapped, y=Genes.Detected)) +
    geom_point() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped")
```

## Exonic mapping rate
```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation (Pearson) heatmap of TMM-normalized counts
```{r pearson-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="pearson"))
```

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="spearman"))
```

## MDS plot of TMM-normalized counts
```{r mds-normalized}
mds(normalized_counts, k=length(colnames(normalized_counts)) - 1, d = "cor")
```

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select = order(rowMeans(normalized_counts),decreasing=TRUE)[1:30]
heatmap_fn(log2(normalized_counts[select,]+1))
```

## Heatmap by concordance correlation coefficient
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075

```{r propcor-heatmap}
propcor = function(x, y) {
    x = log(x + 0.1)
    y = log(y + 0.1)
    num = 2 * cov(x, y)
    denom = var(x) + var(y)
return(num/denom)}

do_propcor = function(x) {
     mat = list()
     for(i in seq_len(ncol(x))) {
         for(j in seq_len(ncol(x))) {
        x2 = x[, i]
        y2 = x[, j]
        mat = c(mat, propcor(x2, y2)) } }
    mat = unlist(mat)
    mat = matrix(mat, ncol(x), ncol(x))
    colnames(mat) = colnames(x)
    rownames(mat) = colnames(x)
    return(mat)}

heatmap_fn(do_propcor(normalized_counts))
```

# Differential expression

```{r de-setup}
library(DESeq2)
library(DEGreport)
library(vsn)
```

# Differential expression

## Genes different expressed between 4h and 0h

```{r time-deseq}
summarydata$covariate = paste0(summarydata$time, summarydata$condition, summarydata$genotype)

```

```{r time-deseq2-expression-analysis, results='asis'}
counts <- counts[rowSums(counts>1)>2,]
dds = DESeqDataSetFromMatrix(countData=counts,
    colData=summarydata, design = ~ covariate)
dds = DESeq(dds)
rld = rlog(dds)
```

### Dispersion estimates

```{r time-dispersion-estimate}
plotDispEsts(dds)
```
### PCAplot

```{r time-pca}
plotPCA(rld, "covariate")
```

```{r time-deseq2-handler}
handle_deseq2 = function(dds, summarydata, column, all_combs) {
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse="_vs_")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
.make_binary_matrix = function(ma){
  total_genes = as.data.frame(matrix(nrow=length(unique(unlist(ma))),ncol=length(ma)))
  total_genes[] = 0
  rownames(total_genes) = unique(unlist(ma))
  colnames(total_genes) = names(ma)
  for (comp in names(ma)){
    total_genes[ma[[comp]],comp] = 1
  }
  total_genes$Name = as.factor(rownames(total_genes))
  rownames(total_genes) = 1:nrow(total_genes)
  total_genes = total_genes[,c(4,1,2,3)]
  for (nc in 2:4)
    total_genes[,nc] <- as.integer(total_genes[,nc])
  total_genes  
}

```

### MA-plots

```{r time-DESeq-output, results='asis'}
comps = list(wt=c("4hwtnon", "0hwtnon"),
             cdnon=c("4hcdnon", "0hcdnon"),
             cdsnp=c("4hcdsnp", "0hcdsnp"))
all_results = handle_deseq2(dds, summarydata, "covariate", comps)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  DESeq2::plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

### Volcano-plots

```{r time-DESeq-volcano}
for(i in seq(length(all_results))) {
  stats = as.data.frame(all_results[[i]][,c(2,6)])
  p = volcano_density_plot(stats, title=names(all_results)[i], lfc.cutoff=1.5)
  print(p)
}
```

### DEGreport

```{r time-get-groups}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp,"_vs_"))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[2]]
  list(g1,g2)
}
```

### Mean-vs-Variation

```{r time-DEGreport-MV}
plots = list()
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], "covariate")
  plots[[i]] = degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, counts(dds,normalized=TRUE)) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

```{r time-DESeq-tables, results='asis'}
ma = list()
for(i in seq(length(all_results))) {
  cat(paste("Lowest adjusted p-value hits for", names(all_results)[i]))
  out_df = all_results[[i]]
  genes = as.data.frame(out_df) %>% mutate(ensembl=rownames(out_df)) %>% filter(abs(log2FoldChange)>2 & padj<0.01)
  ma[[paste0("c",names(all_results)[i])]] = genes$ensembl
  
  print(knitr::kable(head(out_df)))
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n")
}
```

```{r time-upset}

binary_ma=.make_binary_matrix(ma)
upset(binary_ma, sets = names(ma))
```

## Genes different expressed between groups at 0h

### MA-plots

```{r group-DESeq-output, results='asis'}
comps = list(wt=c("0hwtnon", "0hcdnon"),
             cdnon=c("0hwtnon", "0hcdsnp"),
             cdsnp=c("0hcdsnp", "0hcdnon"))
all_results = handle_deseq2(dds, summarydata, "covariate", comps)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  DESeq2::plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

### Mean-vs-Variation

```{r group-DEGreport-MV}
plots = list()
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], "covariate")
  plots[[i]] = degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, counts(dds,normalized=TRUE)) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Differentially expressed genes

```{r group-DESeq-tables, results='asis'}
ma = list()
for(i in seq(length(all_results))) {
  cat(paste("Lowest adjusted p-value hits for", names(all_results)[i]))
  out_df = all_results[[i]]
  genes = as.data.frame(out_df) %>% mutate(ensembl=rownames(out_df)) %>% filter(abs(log2FoldChange)>2 & padj<0.01)
  ma[[paste0("c",names(all_results)[i])]] = genes$ensembl
  
  print(knitr::kable(head(out_df)))
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n")
}
```

```{r group-upset}

binary_ma=.make_binary_matrix(ma)
upset(binary_ma, sets = names(ma))
```

## Genes different expressed between groups at 4h

### MA-plots

```{r group4h-DESeq-output, results='asis'}
comps = list(wt=c("4hwtnon", "4hcdnon"),
             cdnon=c("4hwtnon", "4hcdsnp"),
             cdsnp=c("4hcdsnp", "4hcdnon"))
all_results = handle_deseq2(dds, summarydata, "covariate", comps)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  DESeq2::plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

### Mean-vs-Variation

```{r group4h-DEGreport-MV}
plots = list()
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], "covariate")
  plots[[i]] = degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, counts(dds,normalized=TRUE)) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Differentially expressed genes

```{r group4h-DESeq-tables, results='asis'}
ma = list()
for(i in seq(length(all_results))) {
  cat(paste("Lowest adjusted p-value hits for", names(all_results)[i]))
  out_df = all_results[[i]]
  genes = as.data.frame(out_df) %>% mutate(ensembl=rownames(out_df)) %>% filter(abs(log2FoldChange)>2 & padj<0.01)
  ma[[paste0("c",names(all_results)[i])]] = genes$ensembl
  
  print(knitr::kable(head(out_df)))
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n")
}
```

```{r group4h-upset}

binary_ma=.make_binary_matrix(ma)
upset(binary_ma, sets = names(ma))
```

