---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

Last updated: `r date()`

```{r setup, echo=FALSE}
source("~/repos/myRfunctions/transactions.R")
dn <- "umemori-fgf7ko/report-0505"
files_path = "~/repos/pipelines/umemori-fgf7ko/rerport-0505/"
setwd(files_path)
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
                      eval=TRUE, fig.width= 9,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
```


```{r render,eval=FALSE,cache=FALSE}
render_2_drop("~/repos/pipelines/umemori-fgf7ko/rerport-0505/rnaseq-report.Rmd",dn)
```

# Overview

> Experiment involves mice with knockout of FGF7. They crossed the knockouts with mice that have GFP expression restricted to interneurons and then used these to purify out interneurons (FACs) for RNA-seq.

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
project_summary = "/home/lpantano/orch/scratch/umemori-fgf7ko/fgf7ko/final/2015-05-04_fgf7ko/project-summary.csv"
counts_file = "/home/lpantano/orch/scratch/umemori-fgf7ko/fgf7ko/final/2015-05-04_fgf7ko/combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]
colnames(counts) = gsub(".counts", "", colnames(counts))
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA.rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias")
```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Mapped reads
```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw() +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

## Genomic mapping rate
```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw() +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Unique mapping rate
```{r unique-rate-plot}
dd = data.frame(Name=names(counts), Unique=colSums(counts), Mapped=summarydata[,"Mapped"])
ggplot(dd, aes(x=Name, y=Unique/Mapped)) +
    geom_bar(stat="identity") +
    ylab("unique mapping rate") + xlab("") +
    theme_bw() +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Duplication rate
```{r duplication-rate-plot}
ggplot(summarydata, aes(x=Name, y=Duplication.Rate.of.Mapped)) +
    geom_bar(stat="identity") +
    ylab("duplication mapping rate") + xlab("") +
    theme_bw() +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```


## Number of genes detected
```{r genes-detected-plot}
dd = data.frame(Name=names(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw() +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Features mapping rate
```{r exonic-mapping-plot}
dd = melt(summarydata[,c("Name","Intergenic.Rate","Intronic.Rate","Exonic.Rate")])
ggplot(dd, aes(x=Name, y=value, fill=variable)) +
    geom_bar(stat="identity", position='dodge') +
    theme_bw() +
    scale_fill_brewer(palette = 'Set1') + 
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("mapping rate") + xlab("")
```

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate)) +
    geom_bar(stat="identity") +
    theme_bw() +
    ylim(0,0.01) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

## Estimated fragment length of paired-end reads
```{r fragment-length-plot}
ggplot(summarydata, aes(x=Name, y=Fragment.Length.Mean)) +
    geom_bar(stat="identity") +
    theme_bw() +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("fragment length") + xlab("")
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw() +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw() +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample, colour=sample)) +
    geom_density() +
    theme_bw() +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

```{r filter}
y = DGEList(counts=counts[rowMeans(counts)>5,])
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
```

## Correlation (Pearson) heatmap of TMM-normalized counts
```{r pearson-heatmap-normalized}
heatmap_fn(cor(log2(normalized_counts+1), method="pearson"))
```

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized}
heatmap_fn(cor(log2(normalized_counts+1), method="spearman"))
```

## MDS plot of TMM-normalized counts
```{r mds-normalized}
mds(log2(normalized_counts+1), k=length(colnames(normalized_counts)) - 1)
```

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select = order(rowMeans(counts),decreasing=TRUE)[1:30]
heatmap_fn(counts[select,])
```

# Differential expression

```{r de-setup}
library(DESeq2)
library(DEGreport)
library(vsn)
design = ~ group
condition = "group"
```

Removing KO2-3 sample because it has a high % of reads to intergenic/intronic regions.

```{r deseq2-expression-analysis, results='asis'}
keep = summarydata$Name[c(1:2,4:6)]
counts <- counts[rowSums(counts>0)>2,keep]
dds = DESeqDataSetFromMatrix(countData=counts,
    colData=summarydata[keep,], design = design)
dds = DESeq(dds)
```

## Effect of variance stabilization

```{r deseq-diagnostics, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5), main="log2")
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5), main="rld")
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5), main="vsd")
```

## Dispersion estimates

```{r dispersion-estimate}
plotDispEsts(dds)
```

```{r deseq2-handler}
handle_deseq2 = function(dds, summarydata, column) {
  all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse="vs")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
```

## MA-plots

```{r DESeq-output, results='asis'}
all_results = handle_deseq2(dds, summarydata, condition)
len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

## Volcano-plots

```{r DESeq-volcano}
for(i in seq(length(all_results))) {
  stats = as.data.frame(all_results[[i]][,c(2,6)])
  p = volcano_density_plot(stats, title=names(all_results)[i], lfc.cutoff=1.5)
  print(p)
}
```

## DEGreport

```{r get-groups}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp,"vs"))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[2]]
  list(g1,g2)
}
```

### Pvalues-vs-Mean

We plot some information about how p-values is correlated with the average mean or
the standard desviation. We should see the same distribution for each p-value bin.

```{r DEGreport-M}
plots = list()
scale_factor = round(1/nr * 14)
for(i in seq(length(all_results))) {
  plots[[i]] = degMean(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Mean for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

These distribution is enriched in high p-values, that means that the model cannot explain all variation from samples.

### Pvalues-vs-Variation

```{r DEGreport-V}
plots = list()
for(i in seq(length(all_results))) {
  plots[[i]] = degVar(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Mean-vs-Variation

```{r DEGreport-MV}
plots = list()
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata[keep,], names(all_results)[i], condition)
  plots[[i]] = degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, counts(dds,normalized=TRUE)) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

## Differentially expressed genes

```{r DESeq-tables, results='asis'}
library(org.Mm.eg.db)
library(AnnotationDbi)

get_symbol <- function(genes){
    symbol = AnnotationDbi::select(org.Mm.eg.db, as.character(genes), "SYMBOL",
                                   keytype="ENSEMBL")
    symbol = symbol %>% distinct(ENSEMBL)
    rownames(symbol) = symbol$ENSEMBL
    map = symbol[genes,]
    map$SYMBOL[is.na(map$SYMBOL)] = map$ENSEMBL[is.na(map$SYMBOL)]
    map$SYMBOL
}

i=1

  cat(paste("Lowest adjusted p-value hits for", names(all_results)[i]))
  out_df = all_results[[i]]
  out_df$symbol = get_symbol(rownames(out_df))
  knitr::kable(head(out_df))
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n")

```

FGF7 has a lot of variation and hasn't a corrected p-value because it will probably lead to a false positive result. 

```{r fgf7, results='asis'}

kable(out_df["ENSMUSG00000027208",])
  
```

## plot top DE genes
```{r plot-top}
top = as.data.frame(out_df) %>% mutate(gene=rownames(out_df)) %>% filter(padj<0.1)

dd = melt(as.data.frame(rlogMat[top$gene,]) %>% mutate(symbol=unlist(top$symbol))) %>% mutate(group=summarydata[variable, "group"])

ggplot(dd, aes(x=group, y=value, colour=variable)) +
    geom_point(alpha=0.5) +
    theme_bw() +
    facet_wrap(~symbol,scales = "free_y")
```


## priorization
Rank genes from less variation in  FC to more variation. Good genes are genes with scores < 0.3. There are a lot of variation here.

```{r degreport-rank, results='hide'}
norm_counts = 2^rlogMat[top$gene,]
rank = suppressMessages(DEGreport::degRank(g[[1]], g[[2]], norm_counts, top$log2FoldChange, 400))
rank$symbol = get_symbol(rownames(rank))
```

```{r show-rank}

kable(rank)

```
## X/Y chromosomes
```{r sex bias, cache=TRUE}
library("biomaRt")
mart = useMart("ensembl", dataset="mmusculus_gene_ensembl")

results=getBM(attributes = c("ensembl_gene_id", "chromosome_name", "external_gene_name"),  mart = mart)

x_genes = results %>% filter(chromosome_name=="X") 

mds(rlogMat[ intersect(x_genes$ensembl_gene_id, rownames(rlogMat)), ], k=2)
```

XIST expression

```{r xist}

kable(rlogMat["ENSMUSG00000086503",])

```

# Files

[ko vs wt](`r get_report_links('kovswt.tsv')`)

