---
output:
    knitrBootstrap::bootstrap_document:
        theme: readable
        highlight: zenburn
        theme.chooser: TRUE
        highlight.chooser: TRUE
    html_document:
        toc: true
        highlight: zenburn

---
    
# Report

This report is removing duplicates using position of read 1 and primer of read 2.
Only uniquely mapped reads are annotated with ensembl gene GTF.

## Overview

This report contains the analysis of 6 samples: 3 ZCCHC6 knockdown and 3 controls

[s1](`r get_report_html("~/repos/pipelines/daley_tailseq/reports/zcchc6/sample1_R1_fastqc.html")`)
[s2](`r get_report_html("~/repos/pipelines/daley_tailseq/reports/zcchc6/sample2_R1_fastqc.html")`)
[s3](`r get_report_html("~/repos/pipelines/daley_tailseq/reports/zcchc6/sample3_R1_fastqc.html")`)
[s4](`r get_report_html("~/repos/pipelines/daley_tailseq/reports/zcchc6/sample4_R1_fastqc.html")`)
[s5](`r get_report_html("~/repos/pipelines/daley_tailseq/reports/zcchc6/sample5_R1_fastqc.html")`)
[s6](`r get_report_html("~/repos/pipelines/daley_tailseq/reports/zcchc6/sample6_R1_fastqc.html")`)


## Method

* Map read 1
* QC for read 1
* Detect polyA in read 2: remove first 5 nt TAG at position ~17
* Overlap read 1 with genes and retrieve read-gene information
* Summarize read 1 and read 2 to get information for each gene (counts, polyA, modifications)
As well an extra tuner step is applied to get more precisely polyA fragment.
* Remove false positive polyA. I mapped read 2
to the genome and remove reads that map entirely since we are trying to detect 
post-modifications. Make sense? There were a lot of reads with short polyA
that ended up with long modifications, and mapped perfectly to the genome. 

### Setup: loading packages and data
    
```{r setup,cache=FALSE}
library(knitr)
library(rmarkdown)
library(knitrBootstrap)
library(ggplot2)
library(gplots)
library(reshape)
library(DESeq2)
library(edgeR)
library(genefilter)
library(CHBUtils)
library(gtools)
library(gridExtra)
library(devtools)
library(dplyr)
library(rjson)
library(GGally)
library(biomaRt)
source("~/repos/myRfunctions/transactions.R")
dn <- "daley_tailseq/zcchc6"
root_path<-"~/ody/projects/daley_mRNA_mod/tailseq-zcchc6"
path_files = "~/repos/pipelines/daley_tailseq/reports/zcchc6/"
```

```{r render,eval=FALSE,cache=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
                      cache.path="~/.cache/tailseq-report/html_cached_colon/", eval=TRUE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
render_2_drop("~/repos/pipelines/daley_tailseq/reports/zcchc6/zcchc6.rmd",dn)
```


```{r load,cache=FALSE}
setwd(root_path)
```

```{r functions,cache=F}
source("~/repos/pipelines/daley_tailseq/reports/functions.R")
```

```{r venn-function,cache=FALSE}
get_entrez = function(g){
  require(biomaRt)
  ensembl = useMart('ensembl', dataset = "hsapiens_gene_ensembl")
  df = getBM(attributes=c("ensembl_gene_id", "entrezgene"), filters="ensembl_gene_id", values=g,mart=ensembl)
  return(df[,2])
}


do_df = function(sample, groups){
    data.frame(gene=groups[[sample]],sample=sample,is=1)
}

do_venn_d = function(groups)
{
    n = length(groups)
    all_genes = do.call(rbind, lapply(names(groups), do_df, groups) )
    
    total_genes = all_genes %>% group_by(sample) %>% summarise(total = sum(is)) %>% ungroup()
    all_genes = reshape(all_genes, timevar = "sample",idvar = "gene", direction = "wide")
    all_genes[is.na(all_genes)] = 0
    
    all_genes_sum = all_genes %>% group_by_(.dots = (names(all_genes[,1:n+1]))) %>% summarise(total = n()) %>% ungroup()
    per = apply( all_genes_sum, 1, function(x){
        unlist((x[n+1]/as.vector(total_genes[,2])) * x[1:n] * 100 )
        })
    per[per==0] = NA
    rownames(per) = names(groups)
 
    num = apply( all_genes_sum, 1, function(x){
        unlist(x[n+1] * x[1:n]  )
        })
    num[num==0] = NA
    rownames(num) = names(groups)
 
    
    aheatmap( t(as.matrix(t(per))), annCol = all_genes_sum[,1:n], Colv = NA, Rowv = NA,cellheight = 45, cellwidth = 25, fontsize=9, txt = num )
    
}

```


## polyA in read-2

purple (reads with polyA), blue (reads without polyA) and orange (reads with polyA < 6 nt) are the total amount of reads. red (modifications) are reads with modifications after polyA. And green are the ones without the 5 nt TAG along the read.

These are the stats of read_2 before taking into account the ones where the paired
read_1 maps onto a gene. 

And we applied here another filter to reduce the number of false positive of polyA.
Normally short polyA with long modifications are removed here. We just don't count when read2 maps perfectly to the genome and is at the expected distance from the read1.

```{r ploya,cache=FALSE}
dd = data.frame()
for (file in list.files(paste0(root_path,"/work/"), "[::.:]stat$") ){
    data = read.table(paste0(root_path, "/work/", file) )
    dd = rbind(dd, data)
}
dd$sample = c(rep("s1",5), rep("s2",5), rep("s3",5) , rep("s4",5),
              rep("s5",5) , rep("s6",5))
ggplot( dd, aes( x=sample, y=V2, fill=V1 ) ) +
    geom_bar( stat = 'identity', position = 'dodge') + 
    scale_fill_brewer( palette = "Set1" ) +
    labs( list( y="reads" ) )

```

## reads assigned to genes

The number of reads in genes is quite low. This reduces a lot the coverage. 
The fastqc files show a lot of polyA reads in the first read file. 

```{r reads-genes,cache=FALSE}
dd = data.frame()
for (file in list.files(paste0(root_path,"/work/"), "count_stats.dat$") ){
    data = read.table(paste0(root_path, "/work/", file))
    dd = rbind(dd, data)
}
dd$sample = c(rep("SiCtl-r1",5), rep("SiZ6-1-r1",5), rep("SiCtl-r2",5) , rep("SiZ6-1-r2",5),
              rep("SiCtl-r3",5) , rep("SiZ6-1-r3",5))

ggplot( dd, aes( x=sample, y=V2, fill=V1 ) ) +
    geom_bar( stat = 'identity', position = 'dodge') + 
    scale_fill_brewer( palette = "Set1" ) +
    labs( list( y="reads" ) )

```


We have a lot of unmmaped reads. I tried to do quick inspection of the data and
did the FASTQC analysis for the un-mapped reads of sample 1:

[s1](`r get_report_html("~/repos/pipelines/daley_tailseq/reports/zcchc6/s1_nomapped_fastqc.html")`)

We have a lot of polyA reads (I counted reads with more than 13A and I got 2 mill reads)
Then we have a little bit more than 1 mill reads with the small RNA seq adapter. 

This will add up to 3 mill, so will explain half of the unmapped reads for that sample, still
(I think) a little higher than expected, but makes sense.


## After filtering

Showing sample 1 and 2 only. 

```{r after-filtering,cache=FALSE}
s1_raw = read.table(paste0(root_path,"/work/s1_summary.dat"))
s2_raw = read.table(paste0(root_path,"/work/s2_summary.dat"))

poly_read_summary(s1_raw)
poly_read_summary(s2_raw)
```


```{r polya-filter-200,cache=FALSE}
s3_raw = read.table(paste0(root_path,"/work/s3_summary.dat"))
s4_raw = read.table(paste0(root_path,"/work/s4_summary.dat"))
s5_raw = read.table(paste0(root_path,"/work/s5_summary.dat"))
s6_raw = read.table(paste0(root_path,"/work/s6_summary.dat"))
```


### polyA > 30 and size >= 6 nt

```{r poya-filter-30,cache=FALSE}
s1 = filter_out(s1_raw,30)
s2 = filter_out(s2_raw,30)
s3 = filter_out(s3_raw,30)
s4 = filter_out(s4_raw,30)
s5 = filter_out(s5_raw,30)
s6 = filter_out(s6_raw,30)
genes_s1 = get_set_genes(s1)
genes_s2 = get_set_genes(s2)
genes_s3 = get_set_genes(s3)
genes_s4 = get_set_genes(s4)
genes_s5 = get_set_genes(s5)
genes_s6 = get_set_genes(s6)
#rw = filter_out(rw_raw,30)#genes_rw = get_set_genes(rw)

```

### top family types

Only using the top 1000 expressed genes. There is 25% of ribosomal and histone
genes. 

```{r top40, cache=TRUE}
plot_family(s1)
plot_family(s2)
plot_family(s3)
plot_family(s4)
plot_family(s5)
plot_family(s6)
```

#### Common genes among 4 samples

```{r common-30,cache=FALSE}
type = c("polyA","add-U","add-UU","add-U>3","top 10%","top 5%")
type_name = c("polyA","a","aa","a3","top10","top5")
names(type) = type_name

 void = lapply(type_name, function(x){
     venn(list(SiCtl1=genes_s1[[x]],
               SiCtl2=genes_s3[[x]],
               SiCtl3=genes_s5[[x]]))
     title(type[x])
 })

 void = lapply(type_name, function(x){
     venn(list(SiZ61=genes_s2[[x]],
               SiZ62=genes_s4[[x]],
               SiZ63=genes_s6[[x]]))
     title(type[x])
 })
```


## Summary of samples with polyA>30

```{r polya-summary,cache=FALSE}
dd_list = list("SiCtl-r1"=s1,"SiZ6-1-r1"=s2,"SiCtl-r2"=s3,"SiZ6-1-r2"=s4,"SiCtl-r3"=s5,"SiZ6-1-r3"=s6)
summary_all(dd_list)
frequency_u(dd_list)
```


## Correlation

```{r norm}
groups = c("SiCtl","SiZ6","SiCtl","SiZ6","SiCtl","SiZ6")
design = data.frame(row.names = names(dd_list), condition = groups)
```


### Total gene expression

```{r cor,cache=FALSE}
counts_table = get_counts(dd_list)
ma = get_norm_counts(counts_table, design)
m = log2(ma[[1]]+1)
#ggpairs(m, alpha=0.4,        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(m)))

```

### Genes expression of polyA-transcripts

```{r cor-polyA,cache=FALSE}
m = log2(ma[[2]]+1)
#ggpairs(m, alpha=0.4,diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(m)))

```


###Genes expression of polyA + U modification

```{r cor-a,cache=FALSE}
m = log2(ma[[3]]+1)
#ggpairs(m, alpha=0.4,        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(m)))

```


### Genes expression of polyA + UU modification

```{r cor-aa,cache=FALSE}
m = log2(ma[[4]]+1)
#ggpairs(m, alpha=0.4,        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(m)))
```


### Genes expression of polyA + U>3 modification

```{r cor-c,cache=FALSE}
m = log2(ma[[5]]+1)
#ggpairs(m, alpha=0.4,        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(m)))
```

## Differential expression

I normalized with the factors coming from the total expression of genes.
We are assumming that the total gene expression don't change a lot among conditions.

We detect the gene ZCCHC6 as down-regulated in the knockdown samples.

```{r deg,results='asis'}
deg_list_raw = get_DEG(get_counts( list("SiCtl-r1"=s1_raw,"SiZ6-1-r1"=s2_raw,"SiCtl-r2"=s3_raw,"SiZ6-1-r2"=s4_raw,"SiCtl-r3"=s5_raw,"SiZ6-1-r3"=s6_raw)), design)

total_exp = results(deg_list_raw[[1]], contrast = c("condition", "SiZ6", "SiCtl"))
knitr::kable(total_exp["ENSG00000083223",])
```

In general, it seems there is an up-regulation of polyA genes, and down-regulation 
of polyA-U.. genes. But counts are quite low, so there is not enough power to detect
statistically this trend. Beside dispersion seems weird, but I think is due to
the low coverage.

### Total expression analysis

```{r deg-tables, results='asis'}
deg_list = get_DEG(counts_table, design)

plot_DEG = function(name){
    dse = deg_list[[name]]
    plotDispEsts(dse)
    title(name)
    res = results(dse, contrast = c("condition", "SiZ6", "SiCtl") )
    DESeq::plotMA(as.data.frame(res))
    title(name)
    res$gene = row.names(res)
    res = as.data.frame(res) %>% filter(!is.na(padj)) %>% filter(padj<0.1) %>%
        arrange(padj)
    res = get_description(res)
    knitr::kable(res, row.names = TRUE)
}
lapply(names(deg_list)[1],plot_DEG)
```

### Poly-A expression analysis

There is like a trend in up-regulation of polyA genes, although very low, so 
there is no significant genes.

```{r deg-polya}
lapply(names(deg_list)[2],plot_DEG)
```


### Poly-A + U expression analysis

Counts are very low here, so there is no power for the analysis. But there is 
a general trend to down-regulation of uridylation

```{r deg-polya-u}
lapply(names(deg_list)[3],plot_DEG)
```

### Poly-A + UU expression analysis

```{r deg-polya-uu}
lapply(names(deg_list)[4],plot_DEG)
```
