---
output:
    knitrBootstrap::bootstrap_document:
        theme: readable
        highlight: zenburn
        theme.chooser: TRUE
        highlight.chooser: TRUE
    html_document:
        toc: true
        highlight: zenburn

---
```{r custom-code-2}
source("~/repos/myRfunctions/transactions.R")
dn <- "rrbs/good_samples"
root_path<-"~/orch/scratch/rrbs/good_samples/work/"
path_files = "~/repos/pipeline/baccarelli-rrbs/good_samples"
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, 
                    cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, 
                    error=FALSE,echo=FALSE,
                    eval=TRUE,fig.height=9, fig.width=12,
                    message=FALSE, prompt=TRUE, comment='', fig.cap='', 
                    bootstrap.show.code=FALSE)
options(bitmapType = 'cairo')
```

```{r setup, echo=FALSE,eval=F}
render_2_drop("~/repos/pipelines/baccarelli-rrbs/good_samples/report-9samples.Rmd",dn)
```

```{r packages}
library(tidyr)
library(dplyr)
library(ggplot2)
library(reshape)
library(rmarkdown)
library(knitr)
ggplot <- function(...) ggplot2::ggplot(...) + theme_bw() 
```


## Stats from pilot samples

Number of SNPs that changed CpG methylation sites:

```{r single-sample}
dd = read.table(file.path(root_path, "link", "link_stat.tsv"), header=F, skip=2)
names(dd) = c('sample', '# of snps')
dd

```


Number of SNPs shared among samples.

```{r shared-sample}
dd = read.table(file.path(root_path, "link", "link_shared_stat.tsv"), header=F, skip=2)
names(dd) = c('# of samples', '# of common snps')
dd

```

## example of ASM

```{r heatmap}
dd = read.table(file.path(root_path, 'link', 'chr17_all_example.ma'))
names(dd) = c("snp", "pct_methylation", "sample")
#reshape(metrics_by_snp[,c(1,2,5)], timevar = "snp", idvar = "sample", direction="wide")
ggplot(dd, aes(y=sample, x=snp, fill=pct_methylation)) +
  geom_tile() +
  scale_fill_gradient2(low = "ghostwhite", mid = "grey50", high = "black", midpoint = 0.5) +
  ggtitle("ASM site between SNP(chr17:21225310) and  CpG (chr17:21225340)") +
  theme_gray()
```

## alignment for one sample

```{r fn}
read_sequence = function(fn, labs){
    dd = read.table(fn)
    nt = lapply(as.character(dd$V1), function(x){
        sp = unlist(strsplit(x, split=""))
        names(sp) = labs
        sp
    })
    nt=do.call(rbind,Filter(is.vector,nt))
    
    rownames(nt) = paste0("read",1:nrow(nt))
    nt    
}

.color_pos = function(v, pos){
  cpg = pos$cpg
  snp = pos$snp
  names(v) = as.character(names(v))
  nt_col=c("-"='grey60',"A"='red3', "T"='blue2', "C"='green2', "G"='orange3')
  meth_col=c("-"='grey60',"C"='yellow3', "T"='grey50')
  this = rep('grey60', length(v))
  names(this) = names(v)
  log_in = sapply(snp, function(x){
    if (v[x["pos"]] != x["ref"])
        this[x["pos"]] <<- nt_col[v[x["pos"]]]
  })
  
  log_in = sapply(cpg, function(x){
    this[x] <<- meth_col[v[x]]    
  })
  this
}

```


```{r another-example}
nt = read_sequence(file.path(root_path, 'link', 'B1850_chr17_example.txt'),21225295:21225354)
ys = 1:nrow(nt)  
xs=as.numeric(colnames(nt))

nt = nt[order(nt[,"21225309"]),]
plot(x=c(xs[1], xs[length(xs)]), y=c(ys[1], ys[length(ys)]+4), type = 'n', axes=FALSE,
     xlab="position", ylab="")
log=lapply(ys,function(read){
  text(xs, rep(read, dim(nt)[1]), nt[read,], 
       col=.color_pos( nt[read,], list( 
           cpg=list("21225308","21225339"),
           snp= list( c(ref='G',pos="21225309"))
           )))
  })
legend("topleft", legend = c("methylated C", "alternative allele in SNP"), fill = c("yellow3", "red"), horiz = TRUE, box.lty = 0, border=0)

```

## another ASM with multiple SNPs and CpG sites.

```{r text}

nt = read_sequence(file.path(root_path, 'link', 'B1850_chr1_example.txt'),105640191:105640260)


#nt = nt[order(nt[,"190641180"]),]
ys = 1:nrow(nt)  
xs=as.numeric(colnames(nt))

plot(x=c(xs[1], xs[length(xs)]), y=c(ys[1], ys[length(ys)]+3), type = 'n', axes=FALSE,
     xlab="position", ylab="")
log=lapply(ys,function(read){
  text(xs, rep(read, dim(nt)[1]), nt[read,], 
       col=.color_pos( nt[read,], list( 
           cpg=list("105640200","105640240"),
           snp= list( c(ref='T',pos="105640250"),  c(ref='T',pos="105640211"))
           )))
  })
legend("topleft", legend = c("methylated C", "alternative allele in SNP"), fill = c("yellow3", "green2"), horiz = TRUE, box.lty = 0, border=0)

```



 
Report from Bismark

```{r links, results='asis'}
fns = list_files_with_exts("/home/lpantano/repos/pipeline/baccarelli-rrbs/pilot", "html")

for (fn in fns){
   cat(paste0("[",basename(fn),"](",get_report_html(fn),")\n"))
}

```


