---
output:
    knitrBootstrap::bootstrap_document:
        theme: readable
        highlight: zenburn
        theme.chooser: TRUE
        highlight.chooser: TRUE
    html_document:
        toc: true
        highlight: zenburn

---
    
# Report
    
## Overview

There are 4 samples, `s1` and `s2` get long fragments and `s3` and `s4` get short fragments.
Among the pairs, there are some difference in protocol, like amplification cycles 
and different ligation method (i am guessing from the metadata)

## Method

* Map read 1
* QC for read 1
* Detect polyA in read 2: remove first 5 nt TAG at position ~17
* Overlap read 1 with genes and retrieve read-gene information
* Summarize read 1 and read 2 to get information for each gene (counts, polyA, modifications)
As well an extra tuner step is applied to get more precisely polyA fragment.
* Remove false positive polyA. I mapped read 2
to the genome and remove reads that map entirely since we are trying to detect 
post-modifications. Make sense? There were a lot of reads with short polyA
that ended up with long modifications, and mapped perfectly to the genome. 

### Setup: loading packages and data
    
```{r setup}
library(knitr)
library(rmarkdown)
library(knitrBootstrap)
library(ggplot2)
library(gplots)
library(reshape)
library(DESeq2)
library(edgeR)
library(genefilter)
library(CHBUtils)
library(gtools)
library(gridExtra)
library(devtools)
library(dplyr)
library(rjson)
library(GGally)
source("~/repos/myRfunctions/transactions.R")
dn <- "daley_tailseq"
root_path<-"~/ody/projects/daley_mRNA_mod/tailseq-112014"
```

```{r render,eval=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
                      cache.path="~/.cache/tailseq-report/html_cached/", eval=TRUE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
render_2_drop("~/repos/pipelines/daley_tailseq/reports/tailseq.rmd",dn)
```


```{r load}
setwd(root_path)
```

```{r read-summary}
read_summary = function(dd){
    dd = dd %>% filter(V3 > 1 & !grepl("N", V2) & !is.na(V2))
    dd$type = as.character(dd$V2)
    dd$type[ dd$type!="counts" & dd$type!="polyA" ] = "Mod"
    dd_melt = dd %>% group_by(type) %>% summarise(genes=length(unique(V1)))
    p = ggplot(dd_melt, aes(y=genes, x=type)) + 
        geom_bar( stat = 'identity') +
        ggtitle("Number of genes")
    print(p)
    dd_melt = dd %>% group_by(V1,type) %>% summarise(total=sum(V3))
    p <- ggplot(as.data.frame(dd_melt) , aes(x=type, y=total, fill=type))+
        geom_boxplot() + 
        scale_fill_brewer( palette = "Set1" ) + 
        scale_y_log10() +
        ggtitle("Average expression")
    print(p)
    dd_melt = dd %>% filter(V2=="AA" | V2=="A" | V2=="C") %>% 
              dplyr::select(V1,type=V2) %>% group_by(type) %>% 
              summarise(genes=length(unique(V1)))
    p = ggplot(dd_melt, aes(y=genes, x=type)) + 
        geom_bar( stat = 'identity', positions="dodge" ) +
        ggtitle("Number of genes")
    print(p)
}

get_set_genes = function(dd){
    poly = unique((dd %>% filter(V2=="polyA"))[,1])
    aa = unique((dd %>% filter(V2=="AA"))[,1])
    a = unique((dd %>% filter(V2=="A"))[,1])
    c = unique((dd %>% filter(V2=="C"))[,1])
    mod = unique((dd %>% filter(V2!="counts" & V2!="polyA"))[,1])
    n10 = length( c(aa,a,c) ) * 0.1
    n5 = length( c(aa,a,c) ) * 0.05
    top10 = unique(( dd %>% filter( V2=="A" |  V2=="AA" | V2=="C"  ) %>%
                         arrange( desc(V3) ))[1:n10,1])
    top5 = unique(( dd %>% filter( V2=="A" |  V2=="AA" | V2=="C"  ) %>%
                         arrange( desc(V3) ))[1:n5,1])
    return( list(poly, mod, a, aa, c, top10, top5) )
}

filter_out = function(dd){
    dd = dd %>% filter(V3 > 1 & !grepl("N", V2) & !is.na(V2))
    genes = (dd %>% filter( V2=="polyA" & V3>=30 ))[,1]
    dd[ dd$V1 %in%  genes, ]
}

poly_read_summary = function(dd){
    dd = dd %>% filter(V3 > 1 & !grepl("N", V2) & !is.na(V2))
    dd$type = as.character(dd$V2)
    dd$type[ dd$type!="counts" & dd$type!="polyA" ] = "Mod"
    dd_melt = dd %>% group_by(type) %>% summarise( total=sum(V3) )
    p = ggplot(dd_melt, aes(y=total, x=type)) + 
        geom_bar( stat = 'identity') +
        ggtitle("Number of reads")
    print(p)
    dd_melt = dd %>% filter(V2=="AA" | V2=="A" | V2=="C") %>% 
              dplyr::select(mod=V2, V3) %>% group_by(mod) %>%
              summarise( total=sum(V3) )
    p = ggplot(dd_melt, aes(y=total, x=mod)) + 
        geom_bar(stat="identity", position = "dodge") +
        ggtitle("Number of reads")
    print(p)
}


```


## polyA in read-2

purple (reads with polyA), blue (reads without polyA) and orange (reads with polyA < 6 nt) are the total amount of reads. red (modifications) are reads with modifications after polyA. And green are the ones without the 5 nt TAG along the read.

These are the stats of read_2 before taking into account the ones where the paired
read_1 maps onto a gene. In this case half of the reads_1 map on to genes 
(Only it counts reads that map to only one gene, that reduce the number of reads with assigned genes.)


```{r ploya}
dd = data.frame()
for (file in list.files(paste0(root_path,"/work"), "stat$") ){
    data = read.table(paste0(root_path, "/work/", file) )
    dd = rbind(dd, data)
}
dd$sample = c(rep("s1",5), rep("s2",5), rep("s3",5) , rep("s4",5) )
ggplot( dd, aes( x=sample, y=V2, fill=V1 ) ) +
    geom_bar( stat = 'identity', position = 'dodge') + 
    scale_fill_brewer( palette = "Set1" ) +
    labs( list( y="reads" ) )

```

## After filtering

After that filter, we reduce the total number of reads, but still have a lot.

```{r polya-filter}
s1 = read.table(paste0(root_path,"/work/s1_summary.dat"))
s1 = filter_out(s1)
s2 = read.table(paste0(root_path,"/work/s2_summary.dat"))
s2 = filter_out(s2)
s3 = read.table(paste0(root_path,"/work/s3_summary.dat"))
s3 = filter_out(s3)
s4 = read.table(paste0(root_path,"/work/s4_summary.dat"))
s4 = filter_out(s4)
genes_s1 = get_set_genes(s1)
genes_s2 = get_set_genes(s2)
genes_s3 = get_set_genes(s3)
genes_s4 = get_set_genes(s4)

print("s1")
poly_read_summary(s1)
print("s2")
poly_read_summary(s2)
print("s3")
poly_read_summary(s3)
print("s4")
poly_read_summary(s4)
```


# Comparison

It will show comparison among samples


```{r go}
library(org.Hs.eg.db)
set <- universeGO(org.Hs.egGO,"Homo sapiens")
get_entrez = function(g){
  require(biomaRt)
  ensembl = useMart('ensembl', dataset = "hsapiens_gene_ensembl")
  df = getBM(attributes=c("ensembl_gene_id", "entrezgene"), filters="ensembl_gene_id", values=g,mart=ensembl)
  return(df[,2])
}

run_one_go = function(g, name){
    g = g[!is.na(g)]
    mf <- runGO(g,set,"MF",as.list(GOMFCHILDREN))
    #plotGO(mf, name)
    write.table(mf[[1]], paste0(name, ".txt") )
    #print(head(mf[[1]]))
    return(mf[[1]][1:20,])
}

run_go_enrichment = function(g, sample){
    table = vector("list")
    table[[1]] = run_one_go(g[[1]], paste0(sample,"_mf_polyA"))
    table[[2]] = run_one_go(g[[2]], paste0(sample,'_mf_mod'))
    table[[3]] = run_one_go(g[[3]], paste0(sample,'_mf_a'))
    table[[4]] = run_one_go(g[[4]], paste0(sample,'_mf_aa'))
    table[[5]] = run_one_go(g[[5]], paste0(sample,'_mf_c'))
    table[[6]] = run_one_go(g[[6]], paste0(sample,'_mf_top10'))
    table[[7]] = run_one_go(g[[7]], paste0(sample,'_mf_top5'))
    table
}

```


## Common genes

```{r public-data}
sp<-read.table("~/ody/projects/daley_mRNA_mod/tailseq_GSE51299/SPtable.csv",header=T,sep=",")
library(biomaRt)
mart = useMart("ensembl", dataset="hsapiens_gene_ensembl")
g <- getBM( attributes=c("ensembl_gene_id", "hgnc_symbol") , filters=
"hgnc_symbol"    , values =as.character(sp$Gene.Name) ,mart=mart)
idx<-match(sp$Gene.Name,g$hgnc_symbol)
sp$ens<-g[idx,1]
sp.f<-sp[,c(24,2,3,11,12,13,14,15,16,17,18,19),]
```


```{r common}

type = c("polyA","Modification","add-U","add-UU","add-G","top 10%","top 5%")
type_name = c("polyA","mod","a","aa","c","top10","top5")
void = lapply(1:7, function(x){
    venn(list(s1=genes_s1[[x]],
              s2=genes_s2[[x]],
              s3=genes_s3[[x]],
              s4=genes_s4[[x]]))
    title(type[x])
})

all_genes = rbind(data.frame(gene=genes_s1[[1]],sample="s1",is=1),
      data.frame(gene=genes_s2[[1]],sample="s2",is=1),
      data.frame(gene=genes_s3[[1]],sample="s3",is=1),
      data.frame(gene=genes_s4[[1]],sample="s4",is=1)
      )
all_genes = reshape(all_genes, timevar = "sample",idvar = "gene", direction = "wide")
all_genes[is.na(all_genes)] = 0
detect = rowSums(all_genes[,2:5])
table(detect)
```

### s1/s2 with public data

```{r common-public}
public=sp.f[sp.f[,3]>1,1]
public = public[!is.na(public)]
venn(list(s1=genes_s1[[1]],
              s2=genes_s2[[1]],
              public=public))
    title("with polyA and public data")

public=sp.f[sp.f[,5]>1,1]
public = public[!is.na(public)]
venn(list(s1=genes_s1[[3]],
              s2=genes_s2[[3]],
              public=public))
    title("with polyA + U and public data")


public = sp.f[sp.f[,8]>1,1]
public = public[!is.na(public)]
venn(list(s1=genes_s1[[4]],
              s2=genes_s2[[4]],
              public=public))
    title("with polyA + UU and public data")

public = sp.f[sp.f[,6]>1,1]
public = public[!is.na(public)]
venn(list(s1=genes_s1[[5]],
              s2=genes_s2[[5]],
              public=public))
    title("with polyA + G and public data")

```


## Correlation


```{r cor-funct}
get_counts = function(counts){
    y = DGEList(counts=counts)
    y = calcNormFactors(y)
    cpm(y, normalized.lib.sizes=TRUE, log = TRUE)
}
```


### Total gene expression

```{r cor}
all = rbind( s1 %>% mutate( sample="s1"),
             s2 %>% mutate( sample="s2"),
             s3 %>% mutate( sample="s3"),
             s4 %>% mutate( sample="s4") )
all_melt = reshape(all %>% filter( V2=="counts" ), direction = "wide", idvar = c("V1","V2"),  timevar = "sample")
all_melt[is.na(all_melt)] = 0
counts = get_counts(all_melt[,3:6])
ggpairs(counts, alpha=0.4,
        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(counts)))

```

### Genes expression of polyA-transcripts

```{r cor-polyA}
all_melt = reshape(all %>% filter( V2=="polyA" ), direction = "wide", idvar = c("V1","V2"),  timevar = "sample")
all_melt[is.na(all_melt)] = 0
ggpairs(log2(all_melt[,3:6]+0.1), alpha=0.4,
        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(counts)))

```


### Genes expression of polyA + modification

```{r cor-mod}
all_melt = reshape(as.data.frame(all %>% filter( V2!="counts" & V3!="polya" ) %>% 
                                group_by(V1,sample) %>% summarise( total=sum(V3)))
                   , direction = "wide", idvar = c("V1"),  timevar = "sample")
all_melt[is.na(all_melt)] = 0
counts = get_counts(all_melt[,2:5])
ggpairs(counts, alpha=0.4,
        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(counts)))

```


###Genes expression of polyA + U modification

```{r cor-a}
all_melt = reshape(all %>% filter( V2=="A" ), direction = "wide", idvar = c("V1","V2"),  timevar = "sample")
all_melt[is.na(all_melt)] = 0
counts = get_counts(all_melt[,3:6])
ggpairs(counts, alpha=0.4,
        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(counts)))
```


### Genes expression of polyA + UU modification

```{r cor-aa}
all_melt = reshape(all %>% filter( V2=="AA" ), direction = "wide", idvar = c("V1","V2"),  timevar = "sample")
all_melt[is.na(all_melt)] = 0
counts = get_counts(all_melt[,3:6])
ggpairs(counts, alpha=0.4,
        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(counts)))
```


### Genes expression of polyA + G modification

```{r cor-c}
all_melt = reshape(all %>% filter( V2=="C" ), direction = "wide", idvar = c("V1","V2"),  timevar = "sample")
all_melt[is.na(all_melt)] = 0
counts = get_counts(all_melt[,3:6])
ggpairs(counts, alpha=0.4,
        diag = list(continuous = "density"))
ggheatmap.show(ggheatmap(cor(counts)))

```

## By sample analysis

It included stats for each sample and GO enrichment analysis

### Sample 1

```{r do_go-s1, results='asis',cache=TRUE}
read_summary(s1)
g_ezid = lapply(genes_s1, get_entrez)
res_go = run_go_enrichment(g_ezid, "s1")
```
```{r table-s1, results='asis'}
lapply(1:7, function(x){
    knitr::kable(res_go[[x]], caption=type[x])
})
```


### Sample 2

```{r do_go-s2, results='asis',cache=TRUE}
read_summary(s2)
g_ezid = lapply(genes_s2, get_entrez)
res_go = run_go_enrichment(g_ezid, "s2")
```
```{r table-s2, results='asis'}
lapply(1:7, function(x){
    knitr::kable(res_go[[x]], caption=type[x])
})
```


### Sample 3

```{r do_go-s3, results='asis',cache=TRUE}
read_summary(s3)
g_ezid = lapply(genes_s3, get_entrez)
res_go = run_go_enrichment(g_ezid, "s3")
```
```{r table-s3, results='asis'}
lapply(1:7, function(x){
    knitr::kable(res_go[[x]], caption=type[x])
})
```


### Sample 4

```{r do_go-s4, results='asis',cache=TRUE}
read_summary(s4)
g_ezid = lapply(genes_s4, get_entrez)
res_go = run_go_enrichment(g_ezid, "s4")
```
```{r table-s4, results='asis'}
lapply(1:7, function(x){
    knitr::kable(res_go[[x]], caption=type[x])
})
```


### Files

```{r files_go, results='asis',eval=FALSE}
go_files = lapply(c("s1","s2","s3","s4"), function(s){
    files = lapply(1:7, function(x){
        data.frame(plot=paste0("<a href=",s,"_mf_",type_name[x],".png>","See plot for Molecular function for ",type[x]," in ", s,"<a/>"),
                   table=paste0("<a href=",s,"_mf_",type_name[x],".txt>"," Get table for Molecular function for ",type[x]," in ", s,"<a/>"))
        
    })
    do.call(rbind, files)
})
f = copy_batch_2_drop("txt",dn)
f = copy_batch_2_drop("png",dn)
knitr::kable(do.call(rbind, go_files))
```
