---
  html_document:
    toc: true
    highlight: zenburn
    theme: united
---
```{r custom-code}
source("~/repos/myRfunctions/transactions.R")
dn <- "manus_rnaseq/"
root_path<-"~/orch/groups/"
path_files = "~/repos/pipelines/manus_rnaseq/report"
setwd(path_files)
```


```{r setup, echo=FALSE, eval=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
                      eval=TRUE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
render_2_drop("~/repos/pipelines/manus_rnaseq/report/qc-summary.Rmd",dn)

```
# Overview

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
project_summary = "/home/lpantano/orch/groups/manus-rnaseq/manus-rnaseq/res/2015-02-04_manus-rnaseq/project-summary.csv"

counts_file = "/home/lpantano/orch/groups/manus-rnaseq/manus-rnaseq/res/2015-02-04_manus-rnaseq/combined.counts"

xpress_file = "/home/lpantano/orch/groups/manus-rnaseq/manus-rnaseq/res/2015-02-04_manus-rnaseq/combined.isoform.express_counts"

xpress_tpm_file = "/home/lpantano/orch/groups/manus-rnaseq/manus-rnaseq/res/2015-02-04_manus-rnaseq/combined.isoform.express_tpm"

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = read.table(project_summary, header=TRUE, sep=",")
rownames(summarydata) = summarydata$Name
summarydata = summarydata[order(summarydata$Name),]
row.names(summarydata) = sub("-",".",row.names(summarydata))


counts = read.table(counts_file, header=TRUE, row.names="id")
colnames(counts) = sub(".counts","", colnames(counts))
counts = counts[, order(colnames(counts))]

xpress = read.table(xpress_file, header=TRUE, row.names="id")
colnames(xpress) = sub("_eff.counts","", colnames(xpress))
xpress = xpress[, order(colnames(xpress))]

xpress_tpm = read.table(xpress_tpm_file, header=TRUE, row.names="id")
colnames(xpress_tpm) = sub(".tpm","", colnames(xpress_tpm))
xpress_tpm = xpress_tpm[, order(colnames(xpress_tpm))]

```

# Quality control metrics

## Mapped reads
```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapped reads") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Genomic mapping rate
```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Number of genes detected
```{r genes-detected-plot}
ggplot(summarydata, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) + 
    ylab("genes detected") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Exonic mapping rate
```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("exonic mapping rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("rRNA rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```


## Mapped reads in count matrix
```{r mapped-plot}
dd = data.frame( Counted = colSums(counts), Name = colnames(counts) )
ggplot(dd, aes(x=Name, y=Counted)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapped reads") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Mapped reads in multimapping count matrix
```{r mapped-plot}
dd = data.frame( Counted = colSums(xpress), Name = colnames(xpress) )
ggplot(dd, aes(x=Name, y=Counted)) + 
    geom_bar(stat="identity") +
    theme(axis.text.x = element_text(angle = 90)) +
    ylab("mapped reads") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```


## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"))
```

## Correlation heatmap of raw counts
```{r heatmap-raw}
pheatmap(cor(counts),annotation = summarydata[,"rRNA.rate",drop=F],main = "counts")
pheatmap(cor(xpress),annotation = summarydata[,"rRNA.rate",drop=F],main = "eXpress")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)

y = DGEList(counts=xpress)
y = calcNormFactors(y)
normalized_xpress = cpm(y, normalized.lib.sizes=TRUE)

```

## Spearman Correlation heatmap of TMM-normalized counts
```{r heatmap-normalized}
pheatmap(cor(normalized_counts,method = "spearman"),annotation = summarydata[,c("rRNA.rate"),drop=F],main = "counts")
pheatmap(cor(normalized_xpress,method = "spearman"),annotation = summarydata[,"rRNA.rate",drop=F],main = "eXpress")
```

## MDS plot of TMM-normalized counts
```{r mds-normalized}
mds(normalized_counts, k=length(colnames(normalized_counts)) - 1, condition = summarydata$rRNA.rate, d = "cor") + ggtitle("counts")

mds(normalized_counts, k=length(colnames(normalized_counts)) - 1, d = "cor", condition = summarydata$rRNA.rate) + ggtitle("eXpress")

```

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select_c = order(rowMeans(normalized_counts),decreasing=TRUE)[1:30]
select_e = order(rowMeans(normalized_xpress),decreasing=TRUE)[1:30]

pheatmap(as.matrix(normalized_counts[select_c,]),annotation = summarydata[,"rRNA.rate",drop=F],main = "counts")
pheatmap(as.matrix(normalized_xpress[select_e,]),annotation = summarydata[,"rRNA.rate",drop=F],main = "eXpress")

```

```{r de-setup}
library(DESeq2)
library(DEGreport)
library(vsn)
design = ~ group
condition = "group"
```

# Differential expression

```{r deseq2-expression-analysis, results='asis'}
xpress <- xpress[rowSums(xpress>0)>1,]
dds = DESeqDataSetFromMatrix(countData=xpress,
    colData=summarydata, design = design)
dds = DESeq(dds)

xpress_tpm <- xpress_tpm[rowSums(xpress_tpm>0)>1,]
dds_tpm = DESeqDataSetFromMatrix(countData=round(xpress_tpm,0),
    colData=summarydata, design = design)
dds_tpm = DESeq(dds_tpm)

counts <- counts[rowSums(counts>0)>1,]
dds_c = DESeqDataSetFromMatrix(countData=counts,
    colData=summarydata, design = design)
dds_c = DESeq(dds_c)
```

## Effect of variance stabilization

```{r deseq-diagnostics, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)
normMat = counts(dds, normalized=T)

meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

## Dispersion estimates

```{r dispersion-estimate}
plotDispEsts(dds)
```

```{r deseq2-handler}
handle_deseq2 = function(dds, summarydata, column) {
  all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse="vs")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
```

## MA-plots
  
```{r DESeq-output, results='asis'}
all_results = handle_deseq2(dds, summarydata, condition)
all_results = all_results[1:3]
par(mfrow=c(1,3))
for(i in seq(length(all_results))) {
  plotMA(all_results[[i]])
  title(paste("MAplot ", names(all_results)[i]))
}
```

## Volcano-plots

```{r DESeq-volcano}
par(mfrow=c(1,3))
for(i in seq(length(all_results))) {
  stats = as.data.frame(all_results[[i]][,c(2,6)])
  p = volcano_density_plot(stats, title=names(all_results)[i], lfc.cutoff=1.5)
  print(p)
}
```

## DEGreport

```{r get-groups}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp,"vs"))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[2]]
  list(g1,g2)
}
```

### Pvalues-vs-Mean
We plot some information about how p-values is correlated with the average mean or
the standard desviation. We should see the same distribution for each p-value bin.

```{r DEGreport-M}
for(i in seq(length(all_results))) {
  print(degMean(all_results[[i]]$pvalue, rlogMat) + 
  ggtitle(paste0("Pvalues-vs-Mean for ", names(all_results)[i])))
}
```

## markers

```{r markers-xpress}
genes = c(row.names(xpress)[!grepl("^Tb",row.names(xpress))],"Tb2945","Tb2699")
dd = melt(rlogMat[genes,])
dd$group = summarydata[match(dd$X2,row.names(summarydata)),"group"]
dd$X1 = as.character(dd$X1)
dd$X1[dd$X1=="Tb2945"] = "endA"
dd$X1[dd$X1=="Tb2699"] = "recA"
dd$X1 = factor(dd$X1, levels=c("endA","recA","TBs.idi","TEc.dxs","TispDF","TSc.FPPS","TCtF4aN.CDS","TSr.VO.F219L.T487N","Tt69.Sr.CPR"))


ggplot(dd, aes(y=value, x=X2, color=group)) +
    geom_point() +
    facet_wrap(~X1)
```


```{r markers-xpress-tpm}
rld_t <- rlog(dds_tpm)
rlogMat_tpm <- assay(rld_t)

genes = c(row.names(xpress_tpm)[!grepl("^Tb",row.names(xpress_tpm))],"Tb2945","Tb2699")
dd = melt(rlogMat_tpm[genes,])
dd$group = summarydata[match(dd$X2,row.names(summarydata)),"group"]
dd$X1 = as.character(dd$X1)
dd$X1[dd$X1=="Tb2945"] = "endA"
dd$X1[dd$X1=="Tb2699"] = "recA"
dd$X1 = factor(dd$X1, levels=c("endA","recA","TBs.idi","TEc.dxs","TispDF","TSc.FPPS","TCtF4aN.CDS","TSr.VO.F219L.T487N","Tt69.Sr.CPR"))


ggplot(dd, aes(y=value, x=X2, color=group)) +
    geom_point() +
    facet_wrap(~X1)
```


```{r markers-count}
rld_c <- rlog(dds_c)
rlogMat_c <- assay(rld_c)

genes = c(row.names(counts)[!grepl("^b",row.names(counts))],"b2945","b2699")
dd = melt(rlogMat_c[genes,])
dd$group = summarydata[match(dd$X2,row.names(summarydata)),"group"]
dd$X1 = as.character(dd$X1)
dd$X1[dd$X1=="b2945"] = "endA"
dd$X1[dd$X1=="b2699"] = "recA"
dd$X1 = factor(dd$X1, levels=c("endA","recA","Bs.idi","Ec.dxs","ispDF","Sc.FPPS","CtF4aN.CDS","Sr.VO.F219L.T487N","t69.Sr.CPR"))


ggplot(dd, aes(y=value, x=X2, color=group)) +
    geom_point() +
    facet_wrap(~X1) +
    scale_color_brewer(palette = "Set1")
```

## General trend in low expressed

```{r low-C-M}
idx = rowMeans(normMat) < 100
head(idx)
dd = melt(normMat[idx,c(1:3,10:12)])
dd$group = "M"
dd$group[grepl("W",dd$X2)] = "WT"
ggplot(dd, aes(y=value, x=group)) +
    geom_jitter() 
```



## Differentially expressed genes

```{r DESeq-tables, results='asis'}
for(i in seq(length(all_results))) {
  cat(paste("Lowest adjusted p-value hits for", names(all_results)[i]))
  out_df = all_results[[i]]
  knitr::kable(head(out_df))
  write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
                         quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  cat("\n")
}
```
