---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

Last updated `r date()` by Lorena Pantano

```{r setup-opts, echo=FALSE}
options(bitmapType = "cairo")
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", bootstrap.show.code = FALSE, fig.width = 11,
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

```{r render,echo=F}
library(rmarkdown)
library(knitrBootstrap)
source("~/repos/myRfunctions/transactions.R")
# knit("qc_de_analysis.Rmd")
# render_2_drop("qc_de_analysis.Rmd", dn = "Fichorova_EdgeSeq_miRNA_analysis")
```

```{r setup}
library(knitr)

library(ggplot2)
library(reshape)
library(DESeq2)
library(genefilter)
library(CHBUtils)
library(gtools)
library(gridExtra)
library(devtools)
library(dplyr)
library(pheatmap)

knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", fig.width=9,fig.heigh=6,
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
.cor=function(d){
  as.dist(1-cor(d,method = "spearman"))
}
```


```{r load}
raw = read.table("BWH_3_Plate_parsed_QualityControlled.csv", header=T,sep="\t", row.names=1)
raw = raw[2:nrow(raw),]
raw = raw[rowSums(raw)>0,]
meta = read.table("metadata.csv", header=T,sep="\t")
row.names(meta) = colnames(raw)
meta = meta[,c(2,6,8)]
```


# QC plots of raw data

## Total counts for each sample
```{r total}
df = data.frame(counts=colSums(raw),samples=colnames(raw), treatment=meta$Treatment)
ggplot(df, aes(x=samples,y=counts,fill=treatment)) +
  geom_bar(stat="identity") +
  theme(axis.text=element_text(angle=90,hjust=1))
```

## Clustering of raw counts
```{r raw-mds}
mds(raw,d = "cor",condition = meta$Treatment)
```

## Cumulative curve
```{r raw-cum}
ma_cum = do.call(cbind,lapply(raw,function(x){
  cumsum(sort(x, decreasing = T))/sum(x)
})) %>% as.data.frame()
ma_cum$pos=1:nrow(ma_cum)
ggplot(melt(ma_cum,id.vars = "pos"), aes(x=pos, y=value, group=variable)) +
  geom_point() +
  ylab("pct of total reads") +
  xlab("ranked position of miRNA")
```

## Heatmap of raw counts

below are the boxplot of the raw counts for positve and negative probes.

```{r raw-heatmap}
pheatmap(log2(raw+0.5), annotation = meta, clustering_distance_col=.cor(log2(raw+0.5)), clustering_method = "ward.D2",show_rownames = F)

positive = grepl("POS", rownames(raw))
negative = grepl("ANT", rownames(raw))
hk = grepl("HK", rownames(raw))
mirs = grepl("miR", rownames(raw)) | grepl("let", rownames(raw))

dd = melt(raw %>%  mutate(id=rownames(raw)), id.vars = "id")
dd$group = meta[dd$variable, "Treatment"]

ggplot(dd %>% filter(grepl("POS", id))) + geom_boxplot(aes(y=value,x=variable,fill=group)) + ggtitle("Positive probes")
ggplot(dd %>% filter(grepl("ANT", id))) + geom_boxplot(aes(y=value,x=variable,fill=group)) + ggtitle("Negative probes") +
  theme(axis.text=element_text(angle=90,hjust=1))

```

Only one sample has a lot of noise in the negative probes. Remove sample 55 (  P. bivia  replicate).

# Normalization

Kept miRNAs with average counts > 70 in any group.
```{r deseq2, cache=TRUE}
library(DESeq2)

no_pooled = !grepl("Pool", meta$Experimentdate)
clean_meta = meta
clean_meta$Treatment = gsub(" ", "",clean_meta$Treatment)
group = as.character(unique(clean_meta$Treatment[no_pooled]))
clean_raw = raw[grepl("-",rownames(raw)),]
cpm = clean_raw

keep = lapply(group, function(g){
  .is = clean_meta$Treatment == g
  .count_group = rowMeans(cpm[,.is])
  .count_group > 70
})
idx = apply(do.call(cbind, keep),1,any)
all = do.call(cbind, keep)
colnames(all) = gsub(" ","",group)

# mirs = grepl("miR", rownames(raw_f)) | grepl("let", rownames(raw_f))
raw_f = clean_raw[idx,]
dse = DESeqDataSetFromMatrix(raw_f, meta,~ Treatment)
dse = estimateSizeFactors(dse)
rlg = rlog(dse)
# 
# save(rlg,file="rlg.rda")
# load("rls.rda")
```

Total mirnas `r nrow(raw_f)`

## Distribution
```{r rlg-dist}
df = melt(assay(rlg))
df$group = meta$Treatment[match(df$X2,rownames(meta))]
df$experiment = meta$Experimentdate[match(df$X2,rownames(meta))]
ggplot(df, aes(x=value,group=X2,color=group)) + geom_density() + facet_wrap(~experiment)
```

## Clustering
```{r rlg-mds}
ma = assay(rlg)
mds(ma ,d = "cor",condition = meta$Treatment)
```

## Heatmap
```{r rlg-heatmap}
meta$Total = log10(colSums(raw_f))

pheatmap(ma, annotation = meta, clustering_distance_col=.cor(ma), clustering_method = "ward.D2", show_rownames = FALSE)
# pheatmap(ma[positive,], annotation = meta, clustering_distance_col="correlation", clustering_method = "ward.D2")
```

## Pooled samples

Sorted by average expression.

### All conditions
```{r heatmap-pooled}
pooled = grepl("Pool", meta$Experimentdate)
pool_cpm = ma[, pooled]
pool_meta = meta[pooled,]
pool_sort = pool_cpm[order(rowMeans(pool_cpm),decreasing = TRUE),]

pheatmap(cor(pool_cpm,method = "kendall"),clustering_method = "ward.D2", annotation = pool_meta)
pheatmap(pool_sort, annotation = pool_meta, clustering_distance_col=.cor(pool_sort), clustering_method = "ward.D2", show_rownames = FALSE)
pheatmap(pool_sort, annotation = pool_meta[,"Treatment", drop=F], cluster_rows = FALSE, clustering_distance_col=.cor(pool_sort), clustering_method = "ward.D2", show_rownames = FALSE)
```

### Each condition
```{r heatmap-pooled-condition}
for (cond in unique(pool_meta$Treatment)){
  .cond = grepl(cond, pool_meta$Treatment)
  .pool_cpm = pool_cpm[,.cond]
  .pool_meta = pool_meta[.cond,]
  .pool_sort = .pool_cpm[order(rowMeans(.pool_cpm),decreasing = TRUE),]
  pheatmap(.pool_sort, annotation = .pool_meta, cluster_rows = FALSE, clustering_distance_col=.cor(.pool_sort), clustering_method = "ward.D2", show_rownames = FALSE, main=cond)
  pheatmap(.pool_sort, cluster_rows = FALSE, clustering_distance_col=.cor(.pool_sort), clustering_method = "ward.D2", show_rownames = FALSE, main=cond)
}
```

## Not pooled samples

### All
```{r heatmap-nopooled}
library(corrplot)
pooled = !grepl("Pool", meta$Experimentdate)
pool_cpm = ma[, pooled]
pool_meta = meta[pooled,c("Treatment", "Experimentdate")]
# pool_sort = pool_cpm[order(rowMeans(pool_cpm),decreasing = TRUE),]
pheatmap(pool_cpm, annotation = pool_meta[,"Treatment", drop=F], clustering_distance_col=.cor(pool_cpm), clustering_method = "ward.D2", show_rownames = FALSE)

pheatmap(pool_cpm, annotation = pool_meta, clustering_distance_col=.cor(pool_cpm), clustering_method = "ward.D2", show_rownames = FALSE)
```

Correlation values

```{r cor-nopooled,fig.width=12}
pheatmap(cor(pool_cpm,method = "kendall"),clustering_method = "ward.D2", annotation = pool_meta, annotation_row = pool_meta[,"Treatment",drop=F], main="correlation values between samples")

con_crisp = grepl("Control", pool_meta$Treatment) | grepl("crisp", pool_meta$Treatment)
# corrplot(cor(pool_cpm[,con_crisp], method="kendall"), method = "number")
pheatmap(cor(pool_cpm[,con_crisp],method = "kendall"),clustering_method = "ward.D2", annotation = pool_meta, annotation_row = pool_meta[,"Treatment",drop=F], display_numbers = T,main="correlation values between Control and L.crispatus samples")
```

Only for Controls and LCrispatus

```{r cor-nopooled-zoom}
dd = cor(pool_cpm[,con_crisp], method="kendall")
colnames(dd) = paste0("D",pool_meta[colnames(dd),"Experimentdate"],"_",pool_meta[colnames(dd),"Treatment"],"_",1:nrow(dd))
rownames(dd) = paste0("D",pool_meta[rownames(dd),"Experimentdate"],"_",pool_meta[rownames(dd),"Treatment"],"_",1:nrow(dd))
dd = melt(dd)
ggplot(dd %>% filter(grepl("D7/13/2015_L. crispatus",X1) & grepl("L. crispatus",X2) & !grepl("D7/13/2015",X2)), aes(x=X1,y=X2,fill=1))+
  geom_tile(color="white")+
  scale_fill_continuous(guide=FALSE,low="white",high = "white") +
  geom_text(aes(label=round(value,digits=2),size=3))+
  theme_bw()+
  theme(axis.text.x=element_text(angle=90,hjust=1))+
  labs(list(title="correlation among Lcrispatus in different experimental date",x="",y=""))
  

ggplot(dd %>% filter(grepl("D7/13/2015_L. crispatus",X1) & grepl("D7/13/2015_Control",X2)), aes(x=X1,y=X2,fill=1))+
  geom_tile(color="white")+
  scale_fill_continuous(guide=FALSE,low="white",high = "white") +
  geom_text(aes(label=round(value,digits=2),size=3))+
  theme_bw()+
  theme(axis.text.x=element_text(angle=90,hjust=1))+
  labs(list(title="correlation among Lcrispatus and Controls in same experimental date",x="",y=""))
```

### Each condition
```{r heatmap-nopooled-condition}
library(GGally)
for (cond in unique(pool_meta$Treatment)){
  .cond = grepl(cond, pool_meta$Treatment)
  .pool_cpm = pool_cpm[,.cond]
  .pool_meta = pool_meta[.cond,]
  # .pool_sort = .pool_cpm[order(rowMeans(.pool_cpm),decreasing = TRUE),]
  pheatmap(.pool_cpm, annotation = .pool_meta, clustering_distance_col=.cor(.pool_cpm), clustering_method = "ward.D2", show_rownames = FALSE, main=cond)
  # dd_cor = data.frame(run1=rowMeans(.pool_cpm[,1:3]), run2=rowMeans(.pool_cpm[,4:6]), run3=rowMeans(.pool_cpm[,7:9]))
  # print(ggpairs(.pool_cpm))
  
  pheatmap(.pool_cpm, clustering_distance_col=.cor(.pool_cpm), clustering_method = "ward.D2", show_rownames = FALSE, main=cond)
}
```


# Differential expression
```{r fn}
handle_deseq2 = function(dds, summarydata, column) {
  all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    if (comb[1]=="Control")
      comb = rev(comb)
    contrast_string = paste(comb, collapse="_vs_")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}

plot_MA = function(res){
    for(i in seq(length(res))) {
        DESeq2::plotMA(res[[i]])
        title(paste("MA plot for contrast", names(res)[i]))
    }
}

plot_volcano = function(res){
    for(i in seq(length(res))) {
        stats = as.data.frame(res[[i]][,c(2,6)])
        p = volcano_density_plot(stats, title=names(res)[i], lfc.cutoff=1)
        print(p)
    }
}
plot_gene = function(name, d){
  design = colData(d)
  ggplot(plotCounts(d, name, returnData=TRUE, intgroup = "Treatment",transform = TRUE) %>% mutate(run=as.factor(design$Run)), aes(x=Treatment,y=count, color=run)) + 
    geom_point(size=3) +
    ggtitle(name) + 
    scale_color_brewer(palette = "Set1")
}

```


* Not including pooled samples

```{r de}
no_pooled = !grepl("Pool", clean_meta$Experimentdate)
clean_meta$Treatment = relevel(factor(clean_meta$Treatment), ref = "Control")
dse = DESeqDataSetFromMatrix(raw_f[,no_pooled], clean_meta[no_pooled,],~ Treatment)
dse = DESeq(dse)
plotDispEsts(dse)
res = handle_deseq2(dse,colData(dse),"Treatment")
plot_MA(res)
```


```{r nonpooled-de-volcano, results='asis', fig.width=13}
for (i in names(res)){
  cat("\n\n###Top 20 at: ",i,"\n\n")
  print(kable(head(res[[i]],20)))
  # top = (all_results[[i]] %>% mutate(mir=rownames(all_results[[i]])) %>% filter(padj<0.1))[,1:5]
  cat("\n\n") 
  
  stats = as.data.frame(res[[i]][,c(2,6)])
  p = volcano_density_plot(stats, title=i, lfc.cutoff=1)
  # suppressMessages(print(p))
  cat("\n\n")
  
  # .plots = lapply(rownames(res[[i]][1:6,]), plot_gene, dse)
  # do.call(grid.arrange, .plots)
  i = gsub(" ","", i)
  res_tab = cbind(id=rownames(as.data.frame(res[[i]])),as.data.frame(res[[i]]))
  write.table(res_tab,paste0(i,"_mirna_de.xls"),sep="\t",quote=F, row.names=F)
  cat("\n\n") 
  cat("[Download file](",get_report_links(paste0(i,"_mirna_de.xls")),")", sep = "")
  cat("\n\n") 
}

```

## Alternative strategy to detect ON/OFF

I think a valid strategy would be to do the 70 cut off by group.
So I keep mirnas that have average of 70 in any of the groups and then do the 
differential expression analysis. The ones with padj < 0.05 are good candidates 
to on/off assuming 70 as the cutoff for expression/non expression.


```{r on0ff-deseq2, results='asis'}

onoff_de = c()
for (i in names(res)){
  if (grepl("Control", i)){
    .g = unlist(strsplit(i, "_"))
    .g1 = match(.g[1], colnames(all))
    .g2 = match(.g[3], colnames(all))
    .onoff = row.names(res[[i]]) %in% rownames(all[rowSums(all[,c(.g1,.g2)]) == 1 & all[,.g2]==0,])
    .res = res[[i]][.onoff,]
    .res = .res[.res$padj<0.05,]
    cat("\n\n###",.g[3],". Top 10 (total",nrow(.res),"):\n\n")
    print(kable(head(.res, 10)))
    cat("\n\n")
    # top = (all_results[[i]] %>% mutate(mir=rownames(all_results[[i]])) %>% filter(padj<0.1))[,1:5]
    # .plots = lapply(rownames(.res[1:3,]), plot_gene, dse)
    # do.call(grid.arrange, .plots)
    i = gsub(" ","", i)
    res_tab = cbind(id=rownames(as.data.frame(.res)), as.data.frame(.res))
    write.table(res_tab,paste0(i,"_mirna_onoff_de.xls"),sep="\t",quote=F, row.names=F)
    cat("\n\n") 
    cat("[Download file](",get_report_links(paste0(i,"_mirna_onoff_de.xls")),")", sep = "")
    cat("\n\n") 
    onoff_de = c(onoff_de, rownames(.res[.res$padj<0.05,]))
  }
}

```


Heatmap with all onoff detected:

```{r onoff-heatmap}
onoff_rlg = assay(rlog(dse))
pheatmap(onoff_rlg[unique(onoff_de),], clustering_distance_col=.cor(onoff_rlg), clustering_method = "ward.D2", annotation =clean_meta ,show_rownames = FALSE, main="ON/OFF mirnas")

pheatmap(onoff_rlg[unique(onoff_de),], clustering_distance_col=.cor(onoff_rlg), clustering_method = "ward.D2", annotation =clean_meta[,"Treatment", drop=F] ,show_rownames = FALSE, main="ON/OFF mirnas")

```


# Files requested

## heatmaps of pooled samples

[heatmap of pooled samples](`r get_report_links("figure/heatmap-pooled-2.png")`)

## heatmaps of non-pooled samples

### Each condition
[heatmap of non pooled samples from L.Crispatus](`r get_report_links("figure/heatmap-nopooled-condition-2.png")`)

[heatmap of non pooled samples from P.bivia](`r get_report_links("figure/heatmap-nopooled-condition-4.png")`)

[heatmap of non pooled samples from T.Vag UR1](`r get_report_links("figure/heatmap-nopooled-condition-6.png")`)

[heatmap of non pooled samples from Control](`r get_report_links("figure/heatmap-nopooled-condition-8.png")`)

### All together
[heatmap of non pooled samples all together](`r get_report_links("figure/heatmap-nopooled-2.png")`)

## Differential expression tables and volcano plots
```{r table-diff-genes, results='asis', echo=FALSE}
i = 1
for (fn in list.files(".","mirna_de.xls")){
  cat("\n\n[download table with differential expressed genes of ",gsub("_mirna_de.xls","",fn),
      "](",get_report_links(fn),")",sep="")
  
  cat("\n\n[download volcano plot of ",gsub("_mirna_de.tsv","",fn),
      "](",get_report_links(paste0("figure/nonpooled-de-volcano-",i,".png")),")",sep="")
  
  i = i + 1
}
```

## On/Off miRNAs
```{r table-on-off, results='asis', echo=FALSE}
for (fn in list.files(".","onoff_de.xls")){
  if (grepl("Control", fn))
    cat("\n\n[download table with onn/off genes of ",gsub("_onoff_de.xls","",fn),"](",get_report_links(fn),")",sep="")
}
```

