---
title: "RNA-seq analysis."
output:
  html_document:
    toc: true
    toc_depth: 2
    theme: readable
    highlight: zenburn
---

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="RNAseq "
labPI="Sorger"
lablocation="Longwood"
analystname="Lorena Pantano"
analystemail="lpantano@hsph.harvard.edu"
```

RNA-seq analysis for`r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 


```{r knitrsetup, echo=FALSE}
library(rmarkdown)
library(knitr)
knitr::opts_chunk$set(dev="png",echo=FALSE,
                      highlight=TRUE, autodep=TRUE,
                      comment='', fig.cap='',
                      tidy.opts=list(keep.blank.line=FALSE, width.cutoff=200),
                      cache=TRUE,warning=FALSE,message=FALSE,
                      error=FALSE, fig.width = 12, fig.height = 12)
options(bitmapType = 'cairo')
```

```{r project-options, echo=FALSE}
library(myRfunctions) # devtools::install_github('lpantano/myRfunctions')
path_results = "~/orch/scratch/sorger_rnaseq/files"
dir.create(path_results, showWarnings = FALSE)
```

```{r create-report, echo=FALSE, eval=FALSE}
#render_2_drop("mrna-summary.Rmd", "sorger_rnaseq/time_course")
```

# Overview

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(DESeq2)
library(tximport)
library(logging)
library(gridExtra)
library(clusterProfiler)
library(org.Hs.eg.db)
library(cluster)
library(dplyr)

# library(DT)

basicConfig()
extra_metadata = "sample_metadata_batch.tsv"
project_summary = "project-summary.csv"
counts_file = "combined.counts"
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
summarydata$time = as.factor(summarydata$time)
extra_metadata = read.table(extra_metadata, header=T)
summarydata$batch = as.factor(extra_metadata$batch)
summarydata$number = as.factor(extra_metadata$sample_number)

counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]
colnames(counts) = gsub(".counts", "", colnames(counts))
counts <- counts[rowSums(counts>0)>3,]

row.names(summarydata) = paste0(summarydata$number,"_",
                         summarydata$treatment,"_",
                         summarydata$time,"_", 
                         summarydata$concentration)
summarydata$Name2 = row.names(summarydata)
names(counts) = row.names(summarydata)

known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads",
    "Median.insert.size", "Mapped.reads.pct", "Total.reads", "sample_number",
    "avg_coverage_per_region", "Mapped.Reads", "sample_id", "group","Name2", "concentration", "number")
metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]

```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, clustering_method = "ward.D2", clustering_distance_cols = "correlation", ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Mapped reads
```{r mapped-plot, eval="Mapped" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name2, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("") 
```

## Genomic mapping rate
```{r mapping-rate-plot, eval="Mapping.Rate" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name2, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

```{r mapping-rate-plot-noqualimap, eval="Mapped.reads.pct" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name2, y=Mapped.reads.pct)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Number of genes detected
```{r genes-detected-plot, eval="Genes.Detected" %in% colnames(summarydata)}
dd = data.frame(Name=colnames(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Gene detection saturation
```{r saturation-plot, eval="Genes.Detected" %in% colnames(summarydata)}
dd = data.frame(Mapped=summarydata$Mapped, Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Mapped, y=Genes.Detected)) +
    geom_point() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped")
```

## Exonic mapping rate
```{r exonic-mapping-plot, eval="Exonic.Rate" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name2, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

## rRNA|mitRNA mapping rate
```{r check-rRNA-eval}
eval_rRNA = "rRNA_rate" %in% colnames(summarydata) & !sum(is.na(summarydata$rRNA_rate)) == nrow(summarydata)
```
```{r rRNA-rate-plot, eval=eval_rRNA}
ggplot(summarydata, aes(x=Name2, y=rRNA_rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("") +
    facet_wrap(~batch, nrow=3)
```

## Estimated fragment length of paired-end reads
```{r fragment-length-plot, eval="Fragment.Length.Mean" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=Name2, y=Fragment.Length.Mean)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("fragment length") + xlab("")
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(as.data.frame(counts))
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized, eval=ncol(counts) > 1}
heatmap_fn(cor(normalized_counts, method="spearman"))
```

## MDS plot
```{r pca, eval=ncol(counts) > 1}
dds = DESeqDataSetFromMatrix(counts, colData=metadata, design=~1)
vst = varianceStabilizingTransformation(dds)
# plotPCA(vst, intgroup=c("group")) + geom_text(aes(label=summarydata$Name),hjust = 0, nudge_x = 0.3)
mds(assay(vst), condition = metadata$batch)
```

## Heatmap by concordance correlation coefficient
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075

```{r propcor-heatmap, eval=ncol(counts) > 1}
propcor = function(x, y) {
    x = log(x + 0.1)
    y = log(y + 0.1)
    num = 2 * cov(x, y)
    denom = var(x) + var(y)
return(num/denom)}

do_propcor = function(x) {
     mat = list()
     for(i in seq_len(ncol(x))) {
         for(j in seq_len(ncol(x))) {
        x2 = x[, i]
        y2 = x[, j]
        mat = c(mat, propcor(x2, y2)) } }
    mat = unlist(mat)
    mat = matrix(mat, ncol(x), ncol(x))
    colnames(mat) = colnames(x)
    rownames(mat) = colnames(x)
    return(mat)}

heatmap_fn(do_propcor(normalized_counts))
```


```{r de-setup}
library(DESeq2)
# library(DEGreport)
# library(vsn)

plot_top = function(dds, res, n=9){
    metadata = data.frame(colData(dds))
    pp = lapply(1:n, function(i){
        gene= row.names(res[order(res$padj),])[i]
        dd = plotCounts(dds, gene,transform = T,
                        intgroup="time", returnData = T)
        dd$treatment = metadata[row.names(dd), "treatment"]
        dd$batch = metadata[row.names(dd), "batch"]
        p=ggplot(dd, aes(x=time,y=count,color=batch,shape=treatment)) +
            geom_point(size=1, alpha=0.5) +
            theme_bw(base_size = 7) + ggtitle(gene)
        p
    })
    n = ceiling(length(pp))
    do.call(grid.arrange,pp)
    # marrangeGrob(pp, ncol=2, nrow=n)
}

```

# Differential expression

```{r DESeq-output, results='asis'}
save_file <- function(dat, fn, basedir="."){
    if (!is.data.frame(dat))
        dat <- as.data.frame(dat)
    tab <- cbind(id=data.frame(id=row.names(dat)), dat)
    write.table(tab, file.path(basedir, fn), quote=F, sep="\t", row.names=F)
}

print_out = function(all_results, rlog, metadata, prefix, FDR = 0.01, FC=2){
  # cat(paste("## Comparison: ", prefix))
  out_df = as.data.frame(all_results)
  out_df = out_df[!is.na(out_df$padj),]
  out_df = out_df[order(out_df$padj),]
  out_df$symbol = convertIDs(rownames(out_df), 
                             "ENSEMBL", "SYMBOL", 
                             org.Hs.eg.db, "useFirst")
  out_df$description = convertIDs(rownames(out_df),
                                  "ENSEMBL", "GENENAME",
                                  org.Hs.eg.db, "useFirst")
  
  cat("\n",paste(capture.output(summary(all_results))[1:8], collapse = "<br>"),"\n")

  cat("\n\n### MA plot plot\n\n")
  DESeq2::plotMA(all_results)
  title(paste("MA plot for contrast", prefix))
  
  cat("\n\n### Volcano plot\n\n")
  stats = as.data.frame(all_results[,c(2,6)])
  volcano_density_plot(stats, title=prefix, lfc.cutoff=1.5)

  cat("\n\n### Heatmap most significand, padj<",FDR,"and log2FC > ",FC,"\n")
  sign = row.names(out_df)[out_df$padj<FDR & abs(out_df$log2FoldChange) > FC]

  if ( length(sign)<2 ){
    cat("Too few genes to plot.")
  }else{
    pheatmap(rlog[sign, ], show_rownames = F, clustering_distance_rows = "correlation", clustering_method = "ward.D2", annotation_col = metadata)
  }
  cat("\n")
  
  cat("\n\n### Top DE genes\n\n")
  print(kable(head(out_df, 10)))
  fn = paste(prefix, ".tsv", sep="")
  save_file(out_df, fn, path_results)
  cat("\n\nDifferential expression file at: ", fn)
  cat("\n\n")

  cat("\n\n### GO ontology of DE genes (logFC>",FC," and FDR < ",FDR,"):\n\n")
  .res = out_df
  .idx = .res$padj<FDR & abs(.res$log2FoldChange) > FC
  .idx[is.na(.idx)] = FALSE
  .de = out_df$symbol[.idx]
  .accnum = convertIDs(.de, 
                       "SYMBOL", "ENTREZID", 
                       org.Hs.eg.db, "useFirst")
  if (sum(.idx) > 0){
      cat("doing GOenrich with ",sum(.idx), "genes.\n")
      ego <- enrichGO(gene = .accnum[!is.na(.accnum)], 
                      OrgDb = "org.Hs.eg.db", ont = "BP", 
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.01, qvalueCutoff = 0.05, 
                      readable = TRUE)
      
      if (is.data.frame(ego@result)){
          if (nrow(ego@result) > 0){
              fn = paste0(prefix, "_goenrich.tsv")
              save_file(ego@result, fn, path_results)
              print(kable(head(ego@result[,1:7], 20)))
          }
      }
  }
  cat("\n\n")
}
```


## DMSO over time without batch effect correction

```{r deseq2-dmso, results='asis'}
remove = which(summarydata$Name %in%
                 c("24_AGCGATAG-ATAGAGGC", "38_TCCGGAGA-GGCTCTGA", "56_TAATGCGC-AGGCGAAG")
               | summarydata$time=="168")
metadata = summarydata[-remove,]
counts_clean = counts[,metadata$Name2]
cols = c("treatment", "time", "batch")

samples = row.names(metadata[metadata$treatment=="DMSO",])

dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ time)
dds = DESeq(dds, reduced = ~ 1, test = "LRT")
res = results(dds)
rlog = rlog(dds)
```

```{r deseq2-dmso-plot, results='asis'}
print_out(res, 
          assay(rlog), data.frame(colData(dds)), 
          prefix = "dmso_time_wo_batch",
          FC=0, FDR=0.1 )

p = plot_top(dds, res)

```


## DMSO over time with batch effect correction

```{r deseq2-dmso-batch, results='asis'}
dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ batch + time)
dds = DESeq(dds, reduced = ~ batch, test = "LRT")
res = results(dds)
rlog = rlog(dds)
```
```{r deseq2-dmso-batch-plot, results='asis'}

print_out(res, 
          assay(rlog), data.frame(colData(dds)), "dmso_time_w_batch",
          FC=0, FDR=0.05 )


p = plot_top(dds, res)

```


## DMSO vs  Lap over time without batch effect correction

```{r deseq2-lap, results='asis'}
samples = row.names(metadata[(metadata$concentration==3 &
                                  metadata$treatment=="Lap") |
                             metadata$treatment=="DMSO",])

dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ time + treatment + time:treatment)
dds = DESeq(dds, reduced = ~ time + treatment, test = "LRT")
res = results(dds)
rlog = rlog(dds)

print_out(res, 
          assay(rlog), data.frame(colData(dds)),
          prefix = "dmso_lap_time_wo_batch",
          FC=0, FDR=0.1 )

p = plot_top(dds, res, 3)

```

## DMSO vs  Lap over time with batch effect correction

```{r deseq2-lap-batch, results='asis'}

dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ batch + time + treatment + time:treatment)
dds = DESeq(dds, reduced = ~ batch + time + treatment, test = "LRT")
res = results(dds)
rlog = rlog(dds)

print_out(res, 
          assay(rlog), data.frame(colData(dds)), 
          prefix = "dmso_lap_time_w_batch",
          FC=0, FDR=0.05 )


p = plot_top(dds, res)

```


## DMSO vs  Sun over time without batch effect correction

```{r deseq2-sun, results='asis'}
samples = row.names(metadata[(metadata$concentration==3 &
                                  metadata$treatment=="Sun") |
                             metadata$treatment=="DMSO",])

dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ time + treatment + time:treatment)
dds = DESeq(dds, reduced = ~  time + treatment, test = "LRT")
res = results(dds)
rlog = rlog(dds)

print_out(res, 
          assay(rlog), data.frame(colData(dds)), 
          prefix = "dmso_sun_time_wo_batch",
          FC=2, FDR=0.01 )


p = plot_top(dds, res)

```

## DMSO vs  Sun over time with batch effect correction

```{r deseq2-sun-batch, results='asis'}

dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ batch + time + treatment + time:treatment)
dds = DESeq(dds, reduced = ~ batch + time + treatment, test = "LRT")
res = results(dds)
rlog = rlog(dds)

print_out(res, 
          assay(rlog), data.frame(colData(dds)), 
          prefix = "dmso_sun_time_w_batch",
          FC=2, FDR=0.01 )


p = plot_top(dds, res)

```


## DMSO vs  Erl over time without batch effect correction

```{r deseq2-erl, results='asis'}
samples = row.names(metadata[(metadata$concentration==3 &
                                  metadata$treatment=="Erl") |
                             metadata$treatment=="DMSO",])

dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ time + treatment + time:treatment)
dds = DESeq(dds, reduced = ~  time + treatment, test = "LRT")
res = results(dds)
rlog = rlog(dds)

print_out(res, 
          assay(rlog), data.frame(colData(dds)), 
          prefix = "dmso_Erl_time_wo_batch",
          FC=0, FDR=0.05 )


p = plot_top(dds, res)

```

## DMSO vs  Erl over time with batch effect correction

```{r deseq2-erl-batch, results='asis'}

dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ batch + time + treatment + time:treatment)
dds = DESeq(dds, reduced = ~ batch + time + treatment, test = "LRT")
res = results(dds)
rlog = rlog(dds)

print_out(res, 
          assay(rlog), data.frame(colData(dds)), 
          prefix = "dmso_Erl_time_w_batch",
          FC=0, FDR=0.05 )


p = plot_top(dds, res)

```

## DMSO vs  Sora over time without batch effect correction

```{r deseq2-sora, results='asis'}
samples = row.names(metadata[(metadata$concentration==3 &
                                  metadata$treatment=="Sora") |
                             metadata$treatment=="DMSO",])

dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ time + treatment + time:treatment)
dds = DESeq(dds, reduced = ~  time + treatment, test = "LRT")
res = results(dds)
rlog = rlog(dds)

print_out(res, 
          assay(rlog), data.frame(colData(dds)), 
          prefix = "dmso_Sora_time_wo_batch",
          FC=2, FDR=0.01 )


p = plot_top(dds, res)

```

## DMSO vs  Sora over time with batch effect correction

```{r deseq2-sora-batch, results='asis'}

dds = DESeqDataSetFromMatrix(countData=counts_clean[,samples],
                             colData=metadata[samples,cols], 
                             design = ~ batch + time + treatment + time:treatment)
dds = DESeq(dds, reduced = ~ batch + time + treatment, test = "LRT")
res = results(dds)
rlog = rlog(dds)

print_out(res, 
          assay(rlog), data.frame(colData(dds)), 
          prefix = "dmso_Sora_time_w_batch",
          FC=2, FDR=0.01 )


p = plot_top(dds, res)

```


#R Session Info

```{r compress, echo=FALSE, cache=FALSE}
name_res = compress_results(path_results)
```

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```
