---
output:
  knitrBootstrap::bootstrap_document:
    title: "Noga RNAseq"
    theme: readable
    highlight: zenburn
    theme.chooser: FALSE
    highlight.chooser: FALSE
---

```{r knitr,echo=F}
source("~/repos/myRfunctions/transactions.R")
dn <- "noga_tcell_rnaseq"
library(rmarkdown)
library(knitrBootstrap)
library(knitr)
setwd("~/repos/pipelines/noga_rnaseq")
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", fig.width=9,fig.heigh=9, echo=TRUE,
               cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
options(bitmapType = 'cairo')
```

```{r render, eval=FALSE, echo=FALSE}
render_2_drop("~/repos/pipelines/noga_rnaseq/qc-summary.Rmd",dn)
#render("~/repos/pipelines/gregory_tcga/mirna.Rmd")

```


# Overview

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
project_summary = "/home/lpantano/orch/scratch/noga_rnaseq/tcell/final/2015-04-22_tcell/project-summary.csv"
counts_file = "/home/lpantano/orch/scratch/noga_rnaseq/tcell/final/2015-04-22_tcell/combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]
names(counts) = gsub(".counts","",names(counts))
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA.rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias")

summarydata$condition = gsub("[12]","",summarydata$condition)
summarydata$subject = paste0("s", gsub("[AN]","",summarydata$Name))
# summarydata$Name = gsub("[A]","At24",summarydata$Name)
# names(counts) = summarydata$Name = gsub("[A]","At24",names(counts))

```

naive samples are the samples starting with N

24 time point samples are the samples starting with A

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Mapped reads
```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

The depth is low, normally better to have > 10 mill reads.

## Genomic mapping rate
```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Unique mapping rate
```{r unique-rate-plot}
dd = data.frame(Name=names(counts), Unique=colSums(counts), Mapped=summarydata[,"Mapped"])
ggplot(dd, aes(x=Name, y=Unique/Mapped)) +
    geom_bar(stat="identity") +
    ylab("unique mapping rate") + xlab("") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Number of genes detected
```{r genes-detected-plot}
dd = data.frame(Name=names(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Exonic mapping rate
```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

It is a little low, normally the value is above 75%.

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
```{r density-normalized}
melted2 = merge(melted[,c(2,1,3)], summarydata[,c("Name","time")],by=1)
ggplot(melted2, aes(x=count, group=sample, color=time)) +
    geom_density() +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation (Pearson) heatmap of TMM-normalized counts
```{r pearson-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="pearson"))
```

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="spearman"))
```

## MDS plot of TMM-normalized counts
```{r mds-normalized}
mds(normalized_counts, k=length(colnames(normalized_counts)) - 1)
```

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select = order(rowMeans(normalized_counts),decreasing=TRUE)[1:30]
heatmap_fn(normalized_counts[select,])
```

# Normalization

## Difference genes mean

```{r norm-check}
no_zero = rowSums(counts)>1

naive = summarydata[summarydata$time=="N", "Name"]
t24 = summarydata[summarydata$time=="24", "Name"]

y_naive = DGEList(counts=counts[no_zero,naive])
y_naive = calcNormFactors(y_naive)
y_24 = DGEList(counts=counts[no_zero,t24])
y_24 = calcNormFactors(y_24)

norm_naive = cpm(y_naive, normalized.lib.sizes=TRUE, log = TRUE)
norm_24 = cpm(y_24, normalized.lib.sizes=TRUE, log = TRUE)

ratio = rowMeans(norm_24)-rowMeans(norm_naive)
hist(ratio)
```

If we normalize naive and 24h samples separatly and calculate the ratio of the average mean expression
among the two groups, the value is center on 0, meaning that the big majority is not changing, so any standard normalization will be ok.

# Differential expression

```{r de-setup}
library(DESeq2)
library(DEGreport)
library(vsn)
summarydata$group = as.factor(paste0(summarydata$condition,summarydata$time))
design_condition = ~ group
design_time = ~ group + subject
condition = "group"
```

```{r deseq2-expression-analysis, results='asis'}
counts <- counts[rowMeans(counts)>5,]

dds = DESeqDataSetFromMatrix(countData=counts,
    colData=summarydata, design = design_condition)
dds = DESeq(dds)

dds_condition = DESeqDataSetFromMatrix(countData=counts,
    colData=summarydata, design = design_condition)
dds_condition = DESeq(dds_condition)

dds_wt_time = DESeqDataSetFromMatrix(countData=counts[summarydata$Name[summarydata$condition=="WT"]],
    colData=summarydata[summarydata$condition=="WT",], design = design_time)
dds_wt_time = DESeq(dds_wt_time)

dds_ko_time = DESeqDataSetFromMatrix(countData=counts[summarydata$Name[summarydata$condition=="Pgc"]],
    colData=summarydata[summarydata$condition=="Pgc",], design = design_time)
dds_ko_time = DESeq(dds_ko_time)

```

## Effect of variance stabilization

```{r deseq-diagnostics, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

## Dispersion estimates

```{r dispersion-estimate}
plotDispEsts(dds)
```

```{r deseq2-handler}
handle_deseq2 = function(dds, summarydata, column, all_combs) {
  # all_combs = combn(levels(summarydata[,column]), 2, simplify=FALSE)
  all_results = list()
  contrast_strings = list()
  for(comb in all_combs) {
    contrast_string = paste(comb, collapse="vs")
    contrast = c(column, comb)
    res = results(dds, contrast=contrast)
    res = res[order(res$padj),]
    all_results = c(all_results, res)
    contrast_strings = c(contrast_strings, contrast_string)
  }
  names(all_results) = contrast_strings
  return(all_results)
}
```

## MA-plots

```{r DESeq-output, results='asis'}
all_results = handle_deseq2(dds_condition, summarydata, condition,list(c("PgcN","WTN"),c("Pgc24","WT24")))
all_results = c(all_results, handle_deseq2(dds_wt_time, summarydata[summarydata$condition=="WT",], condition, list(c("WT24","WTN"))))
all_results = c(all_results, handle_deseq2(dds_ko_time, summarydata[summarydata$condition=="Pgc",], condition, list(c("Pgc24","PgcN"))))

len = length(all_results)
nr = ceiling( len / 3 )
nc = ceiling( len / nr )
par(mfrow=c(nr,nc))
for(i in seq(length(all_results))) {
  plotMA(all_results[[i]])
  title(paste("MA plot for contrast", names(all_results)[i]))
}
```

## Volcano-plots

```{r DESeq-volcano}
for(i in seq(length(all_results))) {
  stats = as.data.frame(all_results[[i]][,c(2,6)])
  p = volcano_density_plot(stats, title=names(all_results)[i], lfc.cutoff=1.5)
  print(p)
}
```

## DEGreport

```{r get-groups}
get_groups <- function(d, comp, condition)
{
  g <- unlist(strsplit(comp,"vs"))
  g1 <- d$Name[d[, (names(d)==condition)]==g[1]]
  g2 <- d$Name[d[, (names(d)==condition)]==g[2]]
  list(g1,g2)
}
```

### Pvalues-vs-Mean

We plot some information about how p-values is correlated with the average mean or
the standard desviation. We should see the same distribution for each p-value bin.

```{r DEGreport-M}
plots = list()
scale_factor = round(1/nr * 14)
for(i in seq(length(all_results))) {
  plots[[i]] = degMean(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Mean for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Pvalues-vs-Variation

```{r DEGreport-V}
plots = list()
for(i in seq(length(all_results))) {
  plots[[i]] = degVar(all_results[[i]]$pvalue, rlogMat) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Pvalues-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

### Mean-vs-Variation

```{r DEGreport-MV}
plots = list()
for(i in seq(length(all_results))) {
  g <- get_groups(summarydata, names(all_results)[i], condition)
  plots[[i]] = degMV(g[[1]], g[[2]], all_results[[i]]$pvalue, counts(dds,normalized=TRUE)) +
  theme_bw(base_size = scale_factor) +
  ggtitle(paste0("Mean-vs-Variation for ", names(all_results)[i]))
}
do.call(grid.arrange,plots)
```

## Differentially expressed genes

```{r desc-function}
get_description = function(v)
{
    require(biomaRt)
    mart = useMart("ensembl", dataset="mmusculus_gene_ensembl")
    g <- getBM( attributes=c("ensembl_gene_id","external_gene_name","description") , 
                filters="ensembl_gene_id", values =as.character(v) ,mart=mart)
    row.names(g) = g[,1]
    g
}

```

```{r DESeq-tables, results='asis'}
for(i in seq(length(all_results))) {
  cat(paste("Lowest adjusted p-value hits for", names(all_results)[i]))
  out_df = all_results[[i]]
  
  symbol = get_description(row.names(out_df))
  out_df[as.character(symbol$ensembl_gene_id), "symbol"] = symbol$external_gene_name
  out_df[as.character(symbol$ensembl_gene_id), "description"] = symbol$description

  print(knitr::kable(head(out_df)))
  cat("\n")
  #write.table(out_df, file=paste(names(all_results)[i], ".tsv", sep=""),
  #                       quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
  save_file(out_df, paste0(names(all_results)[i], ".tsv"))
  cat("\n")
}
```

## Selected genes

```{r candidates}

genes=c("CD69","IL2Ra","MKI67")

mart = useMart("ensembl", dataset="mmusculus_gene_ensembl")
g <- getBM( attributes=c("ensembl_gene_id","external_gene_name") , 
            filters="external_gene_name", values =genes ,mart=mart)
mat = rlogMat[g$ensembl_gene_id,]
row.names(mat) = g$external_gene_name
heatmap_fn(mat)
```

# Pgc vs WT at 24 h

## Density of log10 TMM-normalized counts of 24h samples
```{r density-normalized-time}
melted2 = merge(melted[,c(2,1,3)], summarydata[,c("Name","time","condition")],by=1) %>%
          filter(time==24)
ggplot(melted2, aes(x=count, group=sample, color=condition)) +
    geom_density() +
    theme_bw(base_size=12) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation among 24h samples
```{r ruvseq, results='asis'}
library(RUVSeq)
dds_time = DESeqDataSetFromMatrix(countData=counts[summarydata$Name[summarydata$time==24]],
    colData=summarydata[summarydata$time==24,], design = design_condition)
dds_time = DESeq(dds_time)

raw = counts(dds_time)
norm = counts(dds_time, normalized=T)

rlog_time = rlog(dds_time)
heatmap_fn(cor(assay(rlog_time)))
```

## Differential expression

We'll remove A10 sample because seems different from the rest. As well, we'll normalize
with RUVSeq a package that try to get all replicate samples the most similar as possible.

```{r remove-de}
keep = c("A12","A14","A4","A6","A8","A9","A1")
mat = as.matrix(round(norm))[rowSums(as.matrix(round(norm)))>1,keep]
set <- newSeqExpressionSet(mat,phenoData =
                               as.data.frame(colData(dds_time)[keep,"group",drop=F]))
differences <- matrix(data=c(1:4, 5:7, -1), byrow=TRUE, nrow=2)
after_ruv <- RUVs(set, row.names(mat), k=1, differences)
    
dds_ruv = DESeqDataSetFromMatrix(countData=raw[row.names(mat),keep],
                                 colData=pData(after_ruv), design = ~ W_1 + group)
dds_ruv = DESeq(dds_ruv)
plotMA(dds_ruv)
```

## table
```{r time24-table, results='asis'}
res = results(dds_ruv, contrast = c("group", "Pgc24", "WT24"))
res = res[order(res$padj),]

cat(paste("Lowest adjusted p-value hits for", "Pgc24" ,"vs" ,"WT24"))

out_df = res
symbol = get_description(row.names(out_df))
out_df[as.character(symbol$ensembl_gene_id), "symbol"] = symbol$external_gene_name
out_df[as.character(symbol$ensembl_gene_id), "description"] = symbol$description

cat("Number of significant genes: ")
print(kable(as.data.frame(table(out_df$padj<0.1))))
cat("\n")
print(knitr::kable(head(out_df)))
cat("\n")
# write.table(out_df, file=paste("Pgc24vsWT24_clean", ".tsv", sep=""),
#            quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
save_file(out_df, paste0("Pgc24vsWT24_clean", ".tsv"))
cat("\n")

```

# time and condition interaction

## Correlation among 24h samples
```{r time_condition-ruvseq, results='asis'}
library(RUVSeq)
keep = setdiff(summarydata$Name, "A10")

dds_int = DESeqDataSetFromMatrix(countData=counts[rowSums(counts)>1, keep],
    colData=summarydata[keep,], design = ~ time*condition)
dds_int = DESeq(dds_int)

raw = counts(dds_int)
norm = counts(dds_int, normalized=T)
```

## Differential expression

We'll remove A10 sample because seems different from the rest. As well, we'll normalize
with RUVSeq a package that try to get all replicate samples the most similar as possible.

```{r time_conditon-de}
set <- newSeqExpressionSet(raw,phenoData =
                               as.data.frame(colData(dds_int)[,c("condition", "time"),drop=F]))
differences <- matrix(data=c(1,6,7,-1,2:5,8,9,14,15,10:13), byrow=TRUE, nrow=4)
after_ruv <- RUVs(set, row.names(raw), k=1, differences)
    
dds_ruv = DESeqDataSetFromMatrix(countData=raw,
                                 colData=pData(after_ruv), design = ~ W_1 + condition*time)
dds_ruv = DESeq(dds_ruv)

res = results(dds_ruv,name = resultsNames(dds_ruv)[5])
res = res[order(res$padj),]
plotMA(res)
```

## table
```{r time-condition-table, results='asis'}
cat(paste("Lowest adjusted p-value hits for", "Pgc24" ,"vs" ,"WT24" ,"over time"))

out_df = res
symbol = get_description(row.names(out_df))
out_df[as.character(symbol$ensembl_gene_id), "symbol"] = symbol$external_gene_name
out_df[as.character(symbol$ensembl_gene_id), "description"] = symbol$description

cat("Number of significant genes: ")
print(kable(as.data.frame(table(out_df$padj<0.1))))
cat("\n")
print(knitr::kable(head(out_df)))
cat("\n")
# write.table(out_df, file=paste("condition_time_interaction", ".tsv", sep=""),
#            quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
save_file(out_df, paste0("condition_time_interaction", ".tsv"))
cat("\n")

```

## figures

Top 20 genes

```{r plot-top, fig.width=10, fig.width=10}
library(reshape)
top20 = row.names(out_df)[1:20]
mat = get_description(top20)
dd = melt(as.data.frame(cbind(rlogMat[top20,],protein=mat[top20,"external_gene_name"])), id.vars="protein")
dd$time = factor(summarydata[dd$variable, "time"], levels=c("N","24"))
dd$condition = summarydata[dd$variable, "condition"]
dd$value = as.numeric(as.character(dd$value))

ggplot(dd, aes(x=time,y=value,colour=condition)) +
    geom_point() +
    theme_bw() +
    scale_color_brewer(palette = "Set1")+
    facet_wrap(~protein, scales = "free")
```



# Files

* [WT 24 vs WT N ](`r get_report_links('WT24vsWTN.tsv')`)
* [Pgc 24 vs Pgc N ](`r get_report_links('Pgc24vsPgcN.tsv')`)
* [Pgc 24 vs WT 24 ](`r get_report_links('Pgc24vsWT24.tsv')`)
* [Pgc N vs WT N ](`r get_report_links('PgcNvsWTN.tsv')`)

* removing one sample: [Pgc 24 vs WT 24 ](`r get_report_links('Pgc24vsWT24_clean.tsv')`)
* [condition*time interaction](r get_report_links('condition_time_interaction.tsv'))
